---
title: "Contamination story"
author: "Sebastian Dall"
date: "2024-02-28"
output: html_document
editor_options: 
  chunk_output_type: console
---



```{r}
pacman::p_load(
  "data.table",
  "ggplot2",
  "stringr",
  "tidyr",
  "dplyr",
  "ggtext",
  "pheatmap",
  "purrr",
  "here",
  "ggnewscale",
  "ggpubr",
  "seqinr"
)
source("../src/constants.R")
source("../src/themes.R")
source("../src/utility.R")
```


```{r}
bin_consensus_a <- fread("../data/anaerobic_digestor/nanomotif/bin-motifs.tsv")
binary_comparison_df_a <- fread("../data/anaerobic_digestor/binnary/anaerobic_digestor_comparison_matrix.tsv")
bin_contamination_a <- fread("../data/anaerobic_digestor/binnary/bin_contamination.tsv")
quality_a <- fread("../data/anaerobic_digestor/mmlong2_lite/results/mmlong2_lite_bins.tsv")
assembly_info_a <- fread("../data/anaerobic_digestor/binnary/contigs_length.tsv")

cov_a <- fread("../data/anaerobic_digestor/mmlong2_lite/tmp/binning/metabat_cov.tsv")

  
  
```

```{r}
bin_consensus_fe <- fread("../data/fecal/nanomotif/bin-motifs.tsv")
binary_comparison_df_fe <- fread("../data/fecal/binnary/fecal_comparison_matrix.tsv")
bin_contamination_fe <- fread("../data/fecal/binnary/bin_contamination.tsv")
quality_fe <- fread("../data/fecal/mmlong2_lite/results/mmlong2_lite_bins.tsv")
assembly_info_fe <- fread("../data/fecal/mmlong2_lite/tmp/flye/assembly_info.txt") %>% 
  rename(
    contig = `#seq_name`
  ) %>% 
  select(contig, length)

cov_fe <- fread("../data/fecal/mmlong2_lite/tmp/binning/metabat_cov.tsv")

```



```{r}
bin_contamination_qc_a <- add_qc_to_bin_contamination(bin_contamination_a, quality_a, assembly_info_a) %>% 
  mutate(sample = "AD")

binary_comparison_df_f_a <- filter_comparison_df(binary_comparison_df_a, assembly_info_a) %>% 
  mutate(sample = "AD")

bin_consensus_a <- prepare_bin_consensus(bin_consensus_a) %>% 
  mutate(sample = "AD")


bin_contamination_qc_fe <- add_qc_to_bin_contamination(bin_contamination_fe, quality_fe, assembly_info_fe) %>% 
  mutate(sample = "FECAL")

binary_comparison_df_f_fe <- filter_comparison_df(binary_comparison_df_fe, assembly_info_fe) %>% 
  mutate(sample = "FECAL")

bin_consensus_fe <- prepare_bin_consensus(bin_consensus_fe) %>% 
  mutate(sample = "FECAL")


```

```{r}
bin_consensus_all <- bind_rows(bin_consensus_a, bin_consensus_fe)

bin_contamination_all <- bind_rows(bin_contamination_qc_a, bin_contamination_qc_fe)


contamination <- bin_contamination_all %>% 
  mutate(is_contamination = TRUE) %>% 
  select(contig, is_contamination, sample)

binary_comparison_all <- bind_rows(binary_comparison_df_f_a, binary_comparison_df_f_fe) %>% 
  left_join(contamination) %>% 
  mutate(is_contamination = ifelse(is.na(is_contamination), FALSE, is_contamination))


```

```{r}
bin_contamination_all %>% 
  group_by(sample, mag_quality) %>%
  summarise(
    n = n()
  )
  
```
```{r}
bin_contamination_all %>% 
  group_by(sample, mag_quality) %>%
  summarise(
    n = n()
  )
  
```


```{r}
plot_methylation_profile <- function(df, bin2see, motifs = NULL, binary = TRUE, annotate = FALSE){
  
  data <- df %>% 
    filter(bin %in% bin2see) %>%
    mutate(
      length_str = paste0(round(length/1000, 2),"kb  "),
      contig = paste(length_str),
      contig = case_when(
        is_contamination ~ paste0('<strong style="color: #b00d02;">', contig, '</strong>'),
        .default = paste0('<span style="color: #808080;">', contig, '</span>')
      )
    ) %>%
    mutate(
      ns = str_extract(motif, "NNN*"),
      ns = str_length(ns),
      motif = str_replace(motif, "NNN*", paste0("(N)<sub>",ns,"</sub>")),
      motif_axis = paste0(
          str_sub(motif, 1, mod_position),
          "<strong>",
          str_sub(motif, mod_position+1, mod_position+1),
          "<sub><sup>",
          map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
          "</sup></sub>",
          "</strong>",
          str_sub(motif, mod_position+2, -1)
        )
    ) %>% 
     
    group_by(motif_mod) %>% 
    filter(sum(methylation_binary, na.rm = TRUE) != 0) %>%
    ungroup()
  
  data_complete <- data %>% 
    select(contig, motif_mod, methylation_binary, mean, length) %>% 
    complete(contig, motif_mod, fill = list(methylation_binary = NA, mean = NA, length = NA))
  
  plot_data <- left_join(data_complete, data %>% select(bin, contig, is_contamination) %>% distinct(contig, .keep_all = TRUE), by = c("contig"))
  
  motif_markdown <- data %>% 
    select(motif_mod, motif_axis) %>% 
    distinct(motif_mod, .keep_all = TRUE)
  
  plot_data <- left_join(plot_data, motif_markdown)
  
  if (!is.null(motifs)) {
      print("not null")
      plot_data <- plot_data %>% 
        filter(motif_mod %in% motifs) 
    }
  
  y_levels = data %>% 
    filter(!is.na(length)) %>% 
    distinct(contig, .keep_all = TRUE) %>%
    arrange(desc(contig_bin),length) %>% 
    select(contig)
  
  
  contamination_plot <- ggplot(plot_data) +
    aes(x = motif_axis, y = factor(contig, levels = y_levels$contig))
  
  
  if (binary) {
    contamination_plot <- contamination_plot +
    geom_tile(aes(fill = methylation_binary), linewidth = 0.2, color = "gray20", )
  } else {
    contamination_plot <- contamination_plot +
      geom_tile(aes(fill = mean), linewidth = 0.2, color = "gray20", )
  }
  
  if (annotate) {
    contamination_plot <- contamination_plot +
      geom_text(aes(label = round(mean,2)))
  }
  
  contamination_plot <- contamination_plot +
    scale_fill_gradient2(limits = c(0, 1), low = "white", high = PLOT_COLORS["dark_blue"], na.value = "white") +
    coord_fixed(ratio = 1, clip = "off") +
    
    geom_text(aes(color = "Contamination"), label = "", alpha = 1) +
    geom_text(aes(color = "No motif observations"), label = "", alpha = 1) +
    
    
    scale_color_manual(values = c("#b00d02", "#808080")) +
    geom_text(data = filter(plot_data, is.na(mean)), label = ".", hjust = 0.5, vjust = 0, size = 7) +
    theme(
      axis.text.y = element_markdown(vjust = 0.5, hjust = 0), # Rotate x-axis labels to 90 degrees
      axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5),
      panel.grid.minor = element_blank(),  # Remove minor gridlines
      panel.background = element_blank(),   # Remove panel background,
      # aspect.ratio = 1
    ) +
    labs(
      x = "Motif",
      y = "",
      fill = "Methylation degree",
      color = ""
    ) +
    guides(
      color = guide_legend(override.aes = list(alpha = 1, fill = c("#b00d02", "black"),  label = c("axis" ,"·"), size = c(2,10)),  direction = "vertical", order = 1)
    )
  
  contamination_plot
}

get_consensus_motifs <- function(df, bin2see) {
  df %>% 
    filter(bin %in% bin2see) %>% 
    pull(motif_mod) %>% 
    unique()
}

get_title <- function(df, bin2see){
  df %>% 
    filter(bin == bin2see) %>% 
    mutate(subtitle = paste0(mag_quality, ", Completeness: ", Completeness, " - Contamination: ", Contamination)) %>% 
    distinct(subtitle)
}
```

```{r}
bin_contamination_all %>% 
  filter(mag_quality == "HQ", binary_methylation_missmatch_score >= 1, length > 20000, n_contaminants <= 2)
```

# Anaerobic digestor
```{r}
plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"), "bin.2.97", motifs = NULL, annotate = T, binary = T)  
plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"), "bin.2.99", motifs = NULL, annotate = T, binary = T)  
plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"), "bin.3.300", motifs = NULL, annotate = T, binary = T) 
plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"), "bin.1.128", motifs = NULL, annotate = T, binary = T) 

plot_methylation_profile(binary_comparison_all %>% filter(sample == "FECAL"), "bin.1.147", motifs = NULL, annotate = F, binary = F) 
```


```{r}
y <- binary_comparison_all %>% filter(sample == "AD") %>% 
  filter(bin == "bin.3.227") %>% 
  filter(length < 13000)
```


```{r}
# motifs = c("AGAGC_a-2", "CAAAAA_a-5", "GATC_a-1", "GTACAY_a-4", "GACAAT_a-4","GCGCT_m-1", "GGGCC_m-3", "AGGC_a-0", "CCNGG_m-1", "CCTTC_m-1", "GACATG_a-3", "GGCGCTRG_m-4", "CACATC_a-3", "CGGAG_a-3", "CGCG_m-0")

motifs = c("CCGCGG_m-1", "AGAGC_a-2", "CCGAGC_a-3", "CAAAAA_a-5", "GATC_a-1", "GTACAY_a-4", "GACAAT_a-4","GCGCT_m-1", "GGGCC_m-3", "AGGC_a-0", "GACATG_a-3","CGGAG_a-3", "CTGCAG_a-4", "GATATC_a-1", "GAGGAG_a-4","GCGATC_a-3", "TTCGAA_a-5")
# motifs = c("AGAGC_a-2", "CAAAAA_a-5", "GATC_a-1", "GTACAY_a-4", "GACAAT_a-4","GCGCT_m-1", "GGGCC_m-3", "AGGC_a-0", "CCNGG_m-1", "CCTTC_m-1", "GACATG_a-3", "GGCGCTRG_m-4", "CACATC_a-3", "CGGAG_a-3", "CGCG_m-0")

motifs = c("CCGCGG_m-1", "AGAGC_a-2", "CCGAGC_a-3", "CAAAAA_a-5", "GATC_a-1", "GTACAY_a-4", "GACAAT_a-4","GCGCT_m-1", "GGGCC_m-3", "AGGC_a-0", "GACATG_a-3","CGGAG_a-3", "CTGCAG_a-4", "GATATC_a-1", "GAGGAG_a-4","GCGATC_a-3", "TTCGAA_a-5")

bins = c("bin.1.257", "bin.1.84")


bins_panel = c(rep("bin.1.257", 18*14), rep("bin.1.257", 18*23))
  #panel=c(rep("Group 1",16*12),rep("Group 2", 12*12), rep("Group 3", 3*12)))
bins = c("bin.1.257", "bin.1.84")


bins_panel = c(rep("bin.1.257", 18*14), rep("bin.1.257", 18*23))
  #panel=c(rep("Group 1",16*12),rep("Group 2", 12*12), rep("Group 3", 3*12)))

bins_q <- quality_a %>%
  mutate(bin = str_remove(bin, "mmlong2_lite.")) %>% 
  filter(bin %in% bins) %>% 
  select(bin, Completeness, Contamination)
  






make_plot <- function(bin2see) {
  data <- binary_comparison_all %>% filter(sample == "AD") %>% 
    filter(bin == bin2see) %>%
make_plot <- function(bin2see) {
  data <- binary_comparison_all %>% filter(sample == "AD") %>% 
    filter(bin == bin2see) %>%
    left_join(bins_q) %>% 
    mutate(
      bin = paste0(bin, ": Comp. ", Completeness, " - Cont. ", sprintf("%.2f", Contamination)),
      length_str = paste0(round(length/1000),"kb"),
      contig = paste(contig,length_str),
      contig = case_when(
        is_contamination ~ paste0('<strong style="color: #b00d02;">', contig, '</strong>'),
        .default = paste0('<span style="color: #808080;">', contig, '</span>')
      )
    ) %>%
    filter(motif_mod %in% motifs)
  
  bin_stats = data$bin[1]
  
  bin_stats = data$bin[1]
  
  data_complete <- data %>% 
    select(contig, motif_mod, methylation_binary, mean, length) %>% 
    complete(contig, motif_mod, fill = list(methylation_binary = NA, mean = NA, length = NA))
  
  plot_data <- left_join(data_complete, data %>% select(bin, contig, is_contamination) %>% distinct(contig, .keep_all = TRUE), by = c("contig"))
  
  motif_markdown <- data %>% 
    select(motif_mod, motif_axis) %>% 
    distinct(motif_mod, .keep_all = TRUE)
  
  plot_data <- left_join(plot_data, motif_markdown)
  
  y_levels = data %>% 
    filter(!is.na(length)) %>% 
    distinct(contig, .keep_all = TRUE) %>%
    arrange(desc(contig_bin),length) %>% 
    select(contig)
  
  
  contamination_plot <- ggplot(plot_data) +
    aes(x = motif_axis, y = factor(contig, levels = y_levels$contig))
  
  
    contamination_plot <- contamination_plot +
      geom_tile(aes(fill = mean), linewidth = 0.2, color = "gray20")
      geom_tile(aes(fill = mean), linewidth = 0.2, color = "gray20")
  
  
  contamination_plot <- contamination_plot +
    scale_fill_gradient2(limits = c(0, 1), low = "white", high = PLOT_COLORS["dark_blue"], na.value = "white", labels = scales::percent, n.breaks = 6, guide = guide_legend(title.position = "top", direction = "horizontal", override.aes = list(alpha = 1))) +
    coord_equal() +
    
    geom_text(aes(color = "Contamination"), label = "", alpha = 1) +
    geom_text(aes(color = "No motif observations"), label = "", alpha = 1) +
    
    
    scale_color_manual(values = c("#b00d02", "black")) +
    geom_text(data = filter(plot_data, is.na(mean)), label = "·", hjust = 0.5, vjust = 0.5, size = 3) +
    theme(
      legend.position = "none",
      axis.text.y = element_markdown(vjust = 0.5, hjust = 0, size = 5), # Rotate x-axis labels to 90 degrees
      axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 5),
      plot.subtitle = element_text(size = 6),
      plot.title = element_text(size = 7),
      strip.text = element_text(size = 6),
      legend.title = element_text(size = 5),
      legend.text = element_text(size = 5),
      legend.key.size = unit(0.3, 'cm'),
      legend.position = "none",
      axis.text.y = element_markdown(vjust = 0.5, hjust = 0, size = 5), # Rotate x-axis labels to 90 degrees
      axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 5),
      plot.subtitle = element_text(size = 6),
      plot.title = element_text(size = 7),
      strip.text = element_text(size = 6),
      legend.title = element_text(size = 5),
      legend.text = element_text(size = 5),
      legend.key.size = unit(0.3, 'cm'),
      panel.grid.minor = element_blank(),  # Remove minor gridlines
      panel.background = element_blank(),   # Remove panel background,
      plot.background = element_blank(),
      plot.background = element_blank(),
    ) +
    labs(
      x = NULL,
      x = NULL,
      y = "",
      fill = "Methylation degree",
      color = "",
      #subtitle = bin_stats
      color = "",
      #subtitle = bin_stats
    ) +
    guides(
      color = guide_legend(override.aes = list(alpha = 1, fill = c("#b00d02", "black"),  label = c("axis" ,"·"), size = c(2)),  direction = "vertical", order = 1)
    ) +
    facet_grid(. ~ bin)
  
  return(contamination_plot)
}


plots_contamination_data = tibble(
  bin = bins
) %>% 
  mutate(
    plt = map(.x = bin, ~make_plot(.x))
  )



plots_contamination_data$plt[[1]] <- plots_contamination_data$plt[[1]] + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + labs(title = "Anaerobic digestor")

plots_contamination_data$plt[[2]] <- plots_contamination_data$plt[[2]] + theme(legend.position = "bottom")


contamination_plt <- plots_contamination_data %>%
  pull(plt) %>%
  patchwork::wrap_plots(ncol = 1)

contamination_plt

# save(contamination_plt, plots_contamination_data, file = "../RData/Contamination_heatmap.RData")
  
# height_hm <- 8
# width_hm <- 6
# plot_name <- "contamination_heatmap_l"
# ggsave(
#   paste0("../figures/", plot_name, ".png"),
#   height = height_hm,
#   width = width_hm,
#   dpi = "retina",
#   bg = "transparent")
```

```{r}
# gc_df_a <- calculateGCContent("../data/anaerobic_digestor/mmlong2_lite/results/mmlong2_lite_assembly.fasta")

# write_delim(gc_df_a, "../data/anaerobic_digestor/gc_content.tsv", delim = "\t")
gc_df_a <- fread("../data/anaerobic_digestor/gc_content.tsv")
# write_delim(gc_df_a, "../data/anaerobic_digestor/gc_content.tsv", delim = "\t")
gc_df_a <- fread("../data/anaerobic_digestor/gc_content.tsv")

# gc_df_fe <- calculateGCContent("../data/fecal/mmlong2_lite/results/mmlong2_lite_assembly.fasta")
# write_delim(gc_df_fe, "../data/fecal/gc_content.tsv", delim = "\t")
# write_delim(gc_df_fe, "../data/fecal/gc_content.tsv", delim = "\t")
```

```{r}
mm_a <- gc_df_a %>% 
  left_join(cov_a %>% rename(contig = contigName)) %>% 
  mutate(sample = "AD") %>% 
  left_join(binary_comparison_all %>% filter(sample == "AD") %>% select(contig, bin, sample, is_contamination, length) %>% distinct(contig, sample, .keep_all = TRUE)) %>% 
  filter(length >= 36900)


bins = c("bin.2.97", "bin.2.99", "bin.3.227")
mm_bin_ad <- mm_a %>% filter(bin %in% bins, sample == "AD")

mm_bin_ad_label <- mm_bin_ad %>% 
  group_by(bin) %>% 
    summarise(
      `1_cov.bam` = mean(`1_cov.bam`),
      gc = mean(gc)
    )


labels <- c("250k", "1000k")
breaks <- c(250, 1000) * 1000

contamination_gc_cov_plt <- mm_a %>% 
  filter(!bin %in% bins) %>% 
ggplot(aes(y = `1_cov.bam`, x = gc * 100, size = length)) +
  geom_point(color = "gray70", alpha = 0.4, pch = 1,) + 
  geom_point(data = mm_bin_ad %>% rename(Contamination = is_contamination), aes(color = Contamination), alpha = 0.7, pch = 16) +
  ggrepel::geom_label_repel(data = mm_bin_ad_label, aes(label = bin, fill = bin), alpha = 0.7, box.padding = 2, size = 1.5, show.legend = FALSE) +
  coord_cartesian(xlim=c(40,70), ylim = c(3,10000)) +
  scale_y_log10() +
  labs(
    y = "Coverage",
    x = "GC (%)",
    title = "Anaerobic Digestor"
  ) +
  scale_color_manual(values = c("#404080",  "#b00d02")) + 
  scale_fill_manual(values = c('white', 'white')) +
  base_plot_theme() +
  theme(
    legend.position = "bottom"
  ) +
  scale_size_continuous(name = "Contig length", breaks = breaks, labels = labels, range = c(0.1, 5)) + 
  guides(color = guide_legend(direction = "horizontal", title.position = "top", title.hjust = 0.1, order = 1), shape = FALSE, size = guide_legend(title.position = "top", title.hjust = 0.2, override.aes = list(color = "gray70", pch = 1))) #+
  # guides(size = guide_legend(title = "Fragment Length"))


# save(contamination_gc_cov_plt, file = "../RData/contamination_gc_cov_data.RData")

height_hm <- 9
width_hm <- 6
plot_name <- "contamination_gc_cov"
ggsave(
  paste0("../figures/", plot_name, ".png"),
  height = height_hm,
  width = width_hm,
  dpi = "retina",
  bg = "transparent")

contamination_gc_cov_plt
```



```{r}
mm_a <- gc_df_a %>% 
  left_join(cov_a %>% rename(contig = contigName)) %>% 
  mutate(sample = "AD")

mm_fe <- gc_df_fe %>% 
  left_join(cov_fe %>% rename(contig = contigName)) %>% 
  mutate(sample = "FECAL")



mm_all <- bind_rows(mm_a, mm_fe) %>% 
  left_join(binary_comparison_all %>% select(contig, bin, sample, is_contamination) %>% distinct(contig, sample, .keep_all = TRUE))

mm_bin_ad <- mm_all %>% filter(bin == "bin.1.32", sample == "AD")


gc <- mm_all %>% 
  filter(bin != "bin.1.32", sample == "AD") %>% 
ggplot(aes(x = `1_cov.bam`, y = gc * 100)) +
  geom_point(color = "gray70", alpha = 0.6) + 
  geom_point(data = mm_bin_ad, aes(color = is_contamination), size=3, alpha = 0.7) +
  coord_cartesian(xlim = c(0,200)) +
  labs(
    x = "Coverage",
    y = "GC (%)"
  ) +
  scale_color_manual(values = c("#404080",  "#b00d02")) + 
  base_plot_theme() +
  theme(
    legend.position = "none",
    aspect.ratio = 1
  )




```


```{r}
bins = c("bin.2.97", "bin.2.99", "bin.3.227")

c1 <- plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"),  bins[1], motifs = NULL, annotate = F, binary = F) +
  labs(x="", subtitle = get_title(bin_contamination_qc_a, bins[1]), title = paste("Anaerobic Digester","-", bins[1]))
c2 <- plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"), "bin.2.99", motifs = NULL, annotate = F, binary = F)+
  labs(x="", subtitle = get_title(bin_contamination_qc_a, bins[2]), title = paste("Anaerobic Digester","-", bins[2]))
c3 <- plot_methylation_profile(binary_comparison_all %>% filter(sample == "AD"), "bin.3.227", motifs = NULL, annotate = F, binary = F)+
  labs(x="", subtitle = get_title(bin_contamination_qc_a, bins[3]), title = paste("Anaerobic Digester","-", bins[3]))
```


```{r}

make_gc_plot <- function(bin2see) {
  
  mm_scores <- gc_df_a %>% 
    left_join(cov_a %>% rename(contig = contigName)) %>% 
    mutate(sample = "AD") %>% 
    left_join(binary_comparison_all %>% filter(sample == "AD") %>% select(contig, bin, sample, is_contamination, length) %>%
    distinct(contig, sample, .keep_all = TRUE))
  
  mm_bin_ad <- mm_scores %>% filter(bin == bin2see, sample == "AD")
  mm_a <- mm_scores %>% 
    filter(length >= 50000)
  
  
  
  
  # Define breaks at exact powers of 10
  breaks <- c(10^(4:6))
  labels <- c(expression(10^4), expression(10^5), expression(10^6))
  
  contamination_gc_cov_plt <- mm_a %>% 
    filter(!bin %in% bins) %>% 
  ggplot(aes(y = `1_cov.bam`, x = gc * 100, size = length)) +
    geom_point(color = "gray70", alpha = 0.6) + 
    geom_point(data = mm_bin_ad %>% rename(Contamination = is_contamination), aes(color = Contamination), alpha = 0.7) +
    coord_cartesian(ylim = c(3,300)) + #xlim=c(30,70)
    scale_y_log10() +
    labs(
      y = "Coverage",
      x = "GC (%)"
    ) +
    scale_color_manual(values = c("#404080",  "#b00d02")) + 
    base_plot_theme() +
    theme(
      legend.position = "bottom"
    ) +
    scale_size_continuous(name = "Contig length", breaks = breaks, labels = labels)
}


gc1 <- make_gc_plot(bins[1])
gc2 <- make_gc_plot(bins[2])
gc3 <- make_gc_plot(bins[3])

c1gc<-ggarrange(c1, gc1, legend = "right")
c2gc<-ggarrange(c2, gc2, legend = "right")
c3gc<-ggarrange(c3, gc3, legend = "right")


save_cgc_plot <- function(plot,name) {
  height_hm <- 8
  width_hm <- 14
  plot_name <- paste0(name)
  ggsave(
    paste0("../figures/", plot_name, ".png"),
    plot,
    height = height_hm,
    width = width_hm,
    dpi = "retina",
    bg = "transparent")
}

save_cgc_plot(c1gc, "bin.2.97")
save_cgc_plot(c2gc, "bin.2.99")
save_cgc_plot(c3gc, "bin.3.227")

```


```{r}

(c1 + c2) / (c3 + contamination_gc_cov_plt)


```

