---
title: "R Notebook"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
editor_options:
  chunk_output_type: inline
---


```{r}
pacman::p_load(
  "data.table",
  "ggplot2",
  "stringr",
  "tidyr",
  "dplyr",
  "ggtext",
  "pheatmap",
  "purrr",
  "here",
  "ggnewscale"
)
source("../src/constants.R")
source("../src/themes.R")
source("../src/utility.R")
```



```{r}
# Prepare motifs
all_motifs_score = fread("../data/motifs-scored_zymo_and_mono.tsv")
direct_label = "direct"
indirect_label = "indirect"
```



```{r}

zymo_motifs <- fread("../data/ZymoHMW/nanomotif/motifs.tsv")[
  , directly := TRUE
]

zymo_motifs_scored <- all_motifs_score[
    sample == "zymo"
  ][
    , sample := NULL
  ]

zymo_motifs <- merge(
  zymo_motifs_scored, 
  zymo_motifs, 
  all.x = TRUE, 
  by=c("contig", "motif", "mod_position", "mod_type"), 
  suffixes = c("", ".drop")
)

zymo_euk <- fread("../data/ZymoHMW/mmlong2_lite/tmp/eukfilt/tiara")
zymo_bin <- fread("../data/ZymoHMW/mmlong2_lite/tmp/binning/contig_bin.tsv", header = FALSE)
colnames(zymo_bin) <-  c("contig", "bin")
zymo_tax <- fread("../data/ZymoHMW/gtdb-tk_mmlong2/gtdbtk.bac120.summary.tsv")

zymo <- merge(zymo_motifs, zymo_euk, by.x = "contig", by.y = "sequence_id", all.x = TRUE)[class_fst_stage == "bacteria"] %>% 
  merge(zymo_bin, by = "contig", all.x = TRUE) %>% 
  merge(zymo_tax, by.x = "bin", by.y = "user_genome", all.x = TRUE)


zymo <- zymo[
    , mean := n_mod / (n_mod + n_nomod)
  ][
    , species := sapply(classification, extract_species_from_gtdbtk)
  ][
    , name := fcase(
      is.na(species), contig,
      !is.na(species), species
    )
  ][
    , name := fcase(
      contig == "contig_105", "*Escherichia coli* Zymo **plasmid I**",
      contig == "contig_89", "*Escherichia coli* Zymo **plasmid II**",
      contig == "contig_66", "*Escherichia coli* Zymo",
      contig == "contig_104", "*Salmonella enterica* **plasmid I**",
      contig == "contig_111", "*Staphylococcus aureus* **plasmid I**",
      contig == "contig_102", "*Staphylococcus aureus* **plasmid II**",
      contig == "contig_93", "*Staphylococcus aureus* **plasmid II**",
      contig != "contig", paste0("*", name, "*")
    )
  ][
    , motif_markdown := paste0(
        str_sub(motif, 1, mod_position),
        "**",
        map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
        str_sub(motif, mod_position+1, mod_position+1),
        "**",
        str_sub(motif, mod_position+2, -1)
    )
  ][
    , simple_count := fcase(
      (n_mod + n_nomod) > 10, "",
      (n_mod + n_nomod) <= 10, as.character(n_mod + n_nomod - 1)
    )
  ][
    , directly := fifelse(is.na(directly), indirect_label, direct_label)
  ]

```




```{r}
# prepare monoculture motifs

gb_motifs_direct <- fread("../data/p_thermoglucosidasius/nanomotif/motifs.tsv")[, directly := TRUE]
gb_motifs <- merge(
  all_motifs_score[sample == "gb"], gb_motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
  , directly := fifelse(is.na(directly), indirect_label, direct_label)
][
  , name := str_replace(contig, "contig", "*Parageobacillus thermoglucosidasius*")
][
  , name := str_replace(name, "_1", "")
][
  , name := str_replace(name, "_2", " **plasmid I**")
][
  , name := str_replace(name, "_3", " **plasmid II**")
][, sample := NULL]


ec_motifs_direct <- fread("../data/e_coli_K12/nanomotif/motifs.tsv")[, directly := TRUE]
ec_motifs <- merge(
  all_motifs_score[sample == "ec"], ec_motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
  , directly := fifelse(is.na(directly), indirect_label, direct_label)
][
  , name := "*Escherichia coli* K12"
][, sample := NULL]


mr_motifs_direct <- fread("../data/m_ruber/nanomotif/motifs.tsv")[, directly := TRUE]
mr_motifs <-  merge(
  all_motifs_score[sample == "mr"], mr_motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
  , directly := fifelse(is.na(directly), indirect_label, direct_label)
][
  , name := "*Meiothermus ruber*"
][, sample := NULL]

columns_to_keep <- colnames(gb_motifs)
monoculture <- rbind(
  gb_motifs, ec_motifs, mr_motifs, zymo[, ..columns_to_keep]
)[, mean := n_mod / (n_mod + n_nomod - 1)][
  , motif_markdown := paste0(
      str_sub(motif, 1, mod_position),
      "<strong>",
      str_sub(motif, mod_position+1, mod_position+1),
      "<sub><sup>",
      map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
      "</sup></sub>",
      "</strong>",
      str_sub(motif, mod_position+2, -1)
  )
][
  , simple_count := fcase(
    (n_mod + n_nomod - 1) > 10, "",
    (n_mod + n_nomod - 1) <= 10, as.character(n_mod+n_nomod-1)
  )
]



```

```{r}
get_axis_order <- function(wide_df) {
  set.seed(1)
  feature_matrix = as.matrix(wide_df[,-1] )
  y_row_order <- pheatmap::pheatmap(as.matrix(feature_matrix), silent = TRUE)$tree_row$order
  y_factor_order <- wide_df$name[y_row_order]
  x_row_order <- pheatmap::pheatmap(t(as.matrix(feature_matrix)), silent = TRUE)$tree_row$order
  x_factor_order <- colnames(wide_df[,-1])[x_row_order]
  return(list(
    y = y_factor_order,
    x = x_factor_order
  ))
}


```



```{r}
monoculture <- monoculture[!is.na(name)][motif_type != "ambiguous"][, max_obs := max(n_mod + n_nomod), by = motif][
  max_obs > 200
]
axis_orders <- get_axis_order(dcast(monoculture, name ~ motif_markdown, value.var = "mean", fill = 0, fun.aggregate = mean))
```


```{r}
plot_data = monoculture
mono_and_zymo_heatmap = ggplot(plot_data) +
  aes(x = factor(motif_markdown, levels = axis_orders[["x"]]), y = factor(name, levels = axis_orders[["y"]])) + 
  geom_tile(aes(fill = directly), linewidth = 0, alpha = 0) + 
  scale_fill_manual(values = c("direct"="forestgreen", "indirect"="#8C228C")) +
  guides(fill = guide_legend(
    title = "Method of detection",
    override.aes = list(alpha = 1))
  ) +
  new_scale("fill") +
  geom_tile(aes(fill = mean), linewidth = 0) +   
  scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "white", high = "forestgreen", na.value = "white") +
  new_scale("fill") +
  geom_tile(data=plot_data[directly == indirect_label], aes(fill = mean), linewidth = 0) + 
  scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "white", high = "#8C228C", na.value = "white") +
  labs(fill = "Directly") + 
  new_scale("fill") +
  geom_tile(aes(fill = mean), linewidth = 0.2, color = "gray20", alpha = 0) +
  scale_fill_gradient2(
    limits = c(0, 1), low = "white", high = "gray20",
    labels = scales::percent
  ) +
  geom_text(data=plot_data[(n_mod + n_nomod - 1)==0], label = ".", hjust=0.5, vjust=0, size=5) +
  coord_fixed(ratio= 1, clip="off") +
  theme(
    axis.text.y = element_markdown(vjust = 0.5, hjust = 0), # Rotate x-axis labels to 90 degrees
    axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5),
    panel.grid.minor = element_blank(),  # Remove minor gridlines
    panel.background = element_blank(),   # Remove panel background,
  ) +
  labs(
    x = "Motif",
    y = "",
    fill = "Methylation degree"
  )

mono_and_zymo_heatmap
```



```{r}
height_hm = 4.5
width_hm = 8
ggsave(
  "../figures/zymo_and_mono_heatmap.png", 
  mono_and_zymo_heatmap, 
  height = height_hm,
  width = width_hm,
  dpi = "retina",
  bg = "white")
ggsave(
  "../figures/zymo_and_mono_heatmap.pdf", 
  mono_and_zymo_heatmap, 
  height = height_hm,
  width = width_hm,
  dpi = "retina",
  bg = "white")
ggsave(
  "../figures/zymo_and_mono_heatmap.tiff", 
  mono_and_zymo_heatmap, 
  height = height_hm,
  width = width_hm,
  dpi = "retina",
  bg = "white")

```





























