---
title: "main_figures"
author: "Jeppe Støtt Bøjer"
date: "2024-02-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Figure requirements

All figures must be cited in sequence within the main article text in the form Fig. 1, Fig. 2.

Figure panels should be prepared at a minimum resolution of 300 dpi and saved at a maximum width of 180 mm.  

Use a 5–7 pt san serif font for standard text labelling and Symbol font for Greek characters. 

Use scale bars, not magnification factors, and include error bars where appropriate. 

Do not flatten labelling or scale/error bars onto images – uneditable or low-resolution images are two of the most common reasons for delay in manuscript preparation.  


```{r}
library(tidyverse)
library(patchwork)
library(grid)
library(scales)
library(gridExtra)
library(gtable)
library(cowplot)
```

# Figure panel 1

### Loading plots for figure panel 1
```{r}
load("/home/bio.aau.dk/wy72yf/Projects/nanomotif-article/figures/benchmark_heatmap_bin_consensus.Rdata")
load("/home/bio.aau.dk/wy72yf/Projects/nanomotif-article/figures/benchmark_heatmap_both.Rdata")
```


```{r}
pacman::p_load(
  "data.table",
  "ggplot2",
  "stringr",
  "tidyr",
  "dplyr",
  "ggtext",
  "pheatmap",
  "purrr",
  "here",
  "ggnewscale",
  "ggpubr",
  "seqinr"
)
source("../src/constants.R")
source("../src/themes.R")
source("../src/utility.R")
```



```{r}
# Prepare motifs
all_motifs_score = fread("../data/motifs-scored_zymo_and_mono.tsv")
direct_label = "Direct"
indirect_label = "Indirect"
direct_color = PLOT_COLORS[["dark_blue"]]
indirect_color = PLOT_COLORS[["dark_green"]]
get_axis_order <- function(wide_df) {
  set.seed(1)
  feature_matrix = as.matrix(wide_df[,-1] )
  y_row_order <- pheatmap::pheatmap(as.matrix(feature_matrix), silent = TRUE)$tree_row$order
  y_factor_order <- wide_df$name[y_row_order]
  x_row_order <- pheatmap::pheatmap(t(as.matrix(feature_matrix)), silent = TRUE)$tree_row$order
  x_factor_order <- colnames(wide_df[,-1])[x_row_order]
  return(list(
    y = y_factor_order,
    x = x_factor_order
  ))
}


```

# Zymo + mono

```{r}

zymo_motifs <- fread("../data/ZymoHMW/nanomotif/motifs.tsv")[
  , directly := TRUE
]
zymo_bin <- fread("../data/ZymoHMW/mmlong2_lite/tmp/binning/contig_bin.tsv", header = FALSE)
colnames(zymo_bin) <-  c("contig", "bin")
zymo_linked <- fread("/home/bio.aau.dk/wy72yf/Projects/restriction-modification-annotation/analysis/ZymoHMW/mtase_linker/nanomotif_assignment_table.tsv")[linked == TRUE][, linked_mtase := TRUE]
zymo_linked <- merge(zymo_linked, zymo_bin)
zymo_motifs_scored <- all_motifs_score[
    sample == "zymo"
  ][
    , sample := NULL
  ]


zymo_motifs <- merge(
  zymo_motifs_scored, 
  zymo_motifs, 
  all.x = TRUE, 
  by=c("contig", "motif", "mod_position", "mod_type"), 
  suffixes = c("", ".drop")
)
zymo_linked[, motif_pos := paste0(contig, motif, mod_type)]
zymo_motifs[, motif_pos := paste0(contig, motif, mod_type)][
  , motif_comp_pos := paste0(contig, motif_complement, mod_type)
  ][
  , linked_mtase := ifelse(
    (motif_pos %in% zymo_linked$motif_pos) | (motif_comp_pos %in% zymo_linked$motif_pos),
    TRUE,
    FALSE
  )
]

zymo_euk <- fread("../data/ZymoHMW/mmlong2_lite/tmp/eukfilt/tiara")

zymo_tax <- fread("../data/ZymoHMW/gtdb-tk_mmlong2/gtdbtk.bac120.summary.tsv")

zymo <- merge(zymo_motifs, zymo_euk, by.x = "contig", by.y = "sequence_id", all.x = TRUE)[class_fst_stage == "bacteria"] %>% 
  merge(zymo_bin, by = "contig", all.x = TRUE) %>% 
  merge(zymo_tax, by.x = "bin", by.y = "user_genome", all.x = TRUE)


zymo <- zymo[
    , mean := n_mod / (n_mod + n_nomod)
  ][
    , species := sapply(classification, extract_species_from_gtdbtk)
  ][
    , name := fcase(
      is.na(species), contig,
      !is.na(species), species
    )
  ][
    , name := fcase(
      contig == "contig_105", "*Escherichia coli* Zymo **plasmid I**",
      contig == "contig_89", "*Escherichia coli* Zymo **plasmid II**",
      contig == "contig_66", "*Escherichia coli* Zymo",
      contig == "contig_104", "*Salmonella enterica* **plasmid I**",
      contig == "contig_111", "*Staphylococcus aureus* **plasmid I**",
      contig == "contig_102", "*Staphylococcus aureus* **plasmid II**",
      contig == "contig_93", "*Staphylococcus aureus* **plasmid II**",
      contig != "contig", paste0("*", name, "*")
    )
  ][
    , motif_markdown := paste0(
        str_sub(motif, 1, mod_position),
        "**",
        map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
        str_sub(motif, mod_position+1, mod_position+1),
        "**",
        str_sub(motif, mod_position+2, -1)
    )
  ][
    , simple_count := fcase(
      (n_mod + n_nomod) > 10, "",
      (n_mod + n_nomod) <= 10, as.character(n_mod + n_nomod - 1)
    )
  ][
    , directly := fifelse(is.na(directly), indirect_label, direct_label)
  ]

zymo_binmotifs <- fread("../data/ZymoHMW/nanomotif/bin-motifs.tsv")
```




```{r}
# prepare monoculture motifs

gb_motifs_direct <- fread("../data/p_thermoglucosidasius/nanomotif/motifs.tsv")[, directly := TRUE]
gb_linked <- fread("/home/bio.aau.dk/wy72yf/Projects/restriction-modification-annotation/analysis/Geobacillus_3/mtase_linker/MTase_assignment_table.tsv")[detected_motif != ""][, linked_mtase := TRUE][, motif := detected_motif]
gb_motifs <- merge(
  all_motifs_score[sample == "gb"], gb_motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
  , directly := fifelse(is.na(directly), indirect_label, direct_label)
][
  , name := str_replace(contig, "contig", "*P. thermoglucosidasius*")
][
  , name := str_replace(name, "_1", "")
][
  , name := str_replace(name, "_2", " **plasmid I**")
][
  , name := str_replace(name, "_3", " **plasmid II**")
][, sample := NULL]

gb_linked[, motif_pos := paste0(contig, motif, mod_type)]
gb_motifs[, motif_pos := paste0(contig, motif, mod_type)][
  , motif_comp_pos := paste0(contig, motif_complement, mod_type)
  ][
  , linked_mtase := ifelse(
    motif_pos %in% gb_linked$motif_pos | motif_comp_pos %in% gb_linked$motif_pos,
    TRUE,
    FALSE
  )
]

ec_motifs_direct <- fread("../data/e_coli_K12/nanomotif/motifs.tsv")[, directly := TRUE]
ec_linked <- fread("/home/bio.aau.dk/wy72yf/Projects/restriction-modification-annotation/analysis/E.coli_K12_stat_3/mtase_linker/MTase_assignment_table.tsv")[detected_motif != ""][, linked_mtase := TRUE][, motif := detected_motif]
ec_motifs <- merge(
  all_motifs_score[sample == "ec"], ec_motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
  , directly := fifelse(is.na(directly), indirect_label, direct_label)
][
  , name := "*Escherichia coli* K12"
][, sample := NULL]
ec_linked[, motif_pos := paste0(contig, motif, mod_type)]
ec_motifs[, motif_pos := paste0(contig, motif, mod_type)][
  , motif_comp_pos := paste0(contig, motif_complement, mod_type)
  ][
  , linked_mtase := ifelse(
    motif_pos %in% ec_linked$motif_pos | motif_comp_pos %in% ec_linked$motif_pos,
    TRUE,
    FALSE
  )
]

mr_motifs_direct <- fread("../data/m_ruber/nanomotif/motifs.tsv")[, directly := TRUE]
mr_linked <- fread("/home/bio.aau.dk/wy72yf/Projects/restriction-modification-annotation/analysis/M.ruber_1/mtase_linker/MTase_assignment_table.tsv")[detected_motif != ""][, linked_mtase := TRUE][, motif := detected_motif]
mr_motifs <-  merge(
  all_motifs_score[sample == "mr"], mr_motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
  , directly := fifelse(is.na(directly), indirect_label, direct_label)
][
  , name := "*Meiothermus ruber*"
][, sample := NULL]
mr_linked[, motif_pos := paste0(contig, motif, mod_type)]
mr_motifs[, motif_pos := paste0(contig, motif, mod_type)][
  , motif_comp_pos := paste0(contig, motif_complement, mod_type)
  ][
  , linked_mtase := ifelse(
    motif_pos %in% mr_linked$motif_pos | motif_comp_pos %in% mr_linked$motif_pos,
    TRUE,
    FALSE
  )
]


columns_to_keep <- colnames(gb_motifs)
monoculture <- rbind(
  gb_motifs, ec_motifs, mr_motifs, zymo[, ..columns_to_keep]
)[, mean := n_mod / (n_mod + n_nomod - 1)][
  , motif_markdown := paste0(
      str_sub(motif, 1, mod_position),
      "<strong>",
      str_sub(motif, mod_position+1, mod_position+1),
      "<sub><sup>",
      map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
      "</sup></sub>",
      "</strong>",
      str_sub(motif, mod_position+2, -1)
  )
][
  , simple_count := fcase(
    (n_mod + n_nomod - 1) > 10, "",
    (n_mod + n_nomod - 1) <= 10, as.character(n_mod+n_nomod-1)
  )
]

monoculture <- monoculture %>%
  mutate(
    motif_markdown = if_else(
      str_detect(motif_markdown, "N{3,}"), # Check for sequences of three or more 'N's
      # Replace the sequence of 'N's with the desired format
      str_replace_all(motif_markdown, "N{3,}", function(x) paste0("(N)<sub>", str_length(x), "</sub>")),
      motif_markdown # Leave the motif_markdown column unchanged if the condition is not met
    )
  )


```


```{r}
monoculture <- monoculture[!is.na(name)][, max_mean := max(mean, na.rm = TRUE), by = motif][
  max_mean > 0.5
][, max_nmod := max(n_mod), by = motif][
  max_nmod > 80
]
axis_orders <- get_axis_order(dcast(monoculture, name ~ motif_markdown, value.var = "mean", fill = 0, fun.aggregate = mean))
```

```{r}
custom_yorder <-rev(c("*Staphylococcus aureus*",  "*Staphylococcus aureus* **plasmid I**",          "*Staphylococcus aureus* **plasmid II**",
"*Meiothermus ruber*",
"*Escherichia coli* Zymo","*Escherichia coli* Zymo **plasmid I**", "*Escherichia coli* Zymo **plasmid II**", 
"*Escherichia coli* K12", 
"*Salmonella enterica*", "*Salmonella enterica* **plasmid I**",
"*P. thermoglucosidasius*",  "*P. thermoglucosidasius* **plasmid I**" ,"*P. thermoglucosidasius* **plasmid II**",
"*Pseudomonas aeruginosa*",                    
"*Enterococcus faecalis*",
"*Bacillus spizizenii*",                        
"*Listeria monocytogenes_B*"))
custom_x_order <- c(axis_orders[["x"]][str_detect(axis_orders[["x"]], "A<sub")], axis_orders[["x"]][str_detect(axis_orders[["x"]], "C<sub")])
```


```{r}
plot_data = monoculture
mono_and_zymo_heatmap = ggplot(plot_data[mean > 0.2]) +
  aes(x = factor(motif_markdown, levels = custom_x_order), y = factor(name, levels = custom_yorder)) + 
  geom_tile(aes(fill = directly), linewidth = 0, alpha = 0) + 
  scale_fill_manual(values = c("Direct"=direct_color, "Indirect"=indirect_color)) +
  guides(fill = guide_legend(
    title = "Method of detection",
    override.aes = list(alpha = 1), direction = "vertical")
  ) +
  new_scale("fill") +
  geom_tile(aes(fill = mean), linewidth = 0) +   
  scale_fill_steps(guide = "none", limits = c(0, 1), low = "white", high = direct_color, na.value = "white", n.breaks = 6) +
  new_scale("fill") +
  geom_tile(data=plot_data[directly == indirect_label & mean > 0.2], aes(fill = mean), linewidth = 0) + 
  scale_fill_steps(guide = "none", limits = c(0, 1), low = "white", high = indirect_color, na.value = "white", n.breaks = 6) +
  labs(fill = "Directly") + 
  new_scale("fill") +
  geom_tile(data= plot_data, aes(fill = mean), linewidth = 0.1, color = "gray20", alpha = 0) +
  scale_fill_steps(
    limits = c(0, 1), low = "white", high = "gray20",
    labels = scales::percent, n.breaks = 6) +
  geom_text(data=plot_data[(n_mod + n_nomod - 1)==0], label = "·", hjust=0.5, vjust=0.5, size=3) +
  geom_text(data=plot_data[linked_mtase == TRUE], label = "+", hjust=0.5, vjust=0.5, size=2, color = "orange") +
  coord_fixed(ratio= 1, clip="off") +
  theme(
    axis.text.y = element_markdown(vjust = 0.5, hjust = 0,size=12), # Rotate x-axis labels to 90 degrees
    axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 12),
    panel.grid.minor = element_blank(),  # Remove minor gridlines
    panel.background = element_blank(),   # Remove panel background,
    legend.position = "top"
  ) +
  labs(
    x = NULL,
    y = NULL,
    title = NULL,
    fill = "Methylation degree"
  ) +
  geom_text(aes(color = "No motif observations"), alpha = 0, label ="") +
  geom_text(aes(color = "Linked MTase"), alpha=0, label = "") +
  scale_color_manual(name = NULL, values = c("No motif observations" = "black", "Linked MTase" = "orange")) +
  guides(color = guide_legend(direction = "vertical", override.aes = list(alpha = 1, label = c("+", "·"), size = c(2, 3)),  fill = c("orange", "black")), fill = guide_legend(title.position = "top", direction = "horizontal", override.aes = list(alpha = 1))) # Adjust this line as needed

mono_and_zymo_heatmap
```





### Adjusting theme and generating a grob object for each ggplot

```{r}
main_theme <- theme(
    #text = element_text(family = "sans", size = 1),
    plot.title = element_text(size = 5, family = "sans"),
    strip.text = element_markdown(size = 5, family = "sans"),
    strip.text.y.right = element_markdown(face = "bold", size = 6, family = "sans"),
    axis.text.y = element_markdown(size=5, family = "sans"), # Rotate x-axis labels to 90 degrees
    axis.text.x = element_markdown(size = 5, family = "sans"),
    axis.title.x = element_text(size = 7, family = "sans"),
    axis.title.y = element_text(size = 7, family = "sans"),
    axis.ticks = element_line(linewidth = 0.1),
    legend.title = element_text(size = 5, family = "sans"),
    legend.text = element_text(size = 5, family = "sans"),
    legend.key.size = unit(0.2, "cm"))




p2 <- bin_consensus_benchmark + main_theme + theme(legend.position = "bottom",
                                                   legend.margin = margin(b = 3, unit = "pt"))
p3 <- mono_and_zymo_heatmap & main_theme & theme(legend.key.size = unit(0.2, "cm"),
                                                 #legend.position = "top",
                                                 #legend.spacing.x = unit(0.01, "cm"),
                                                 legend.justification = c(1, 0),
                                                 legend.box.just = "right",
                                                 legend.margin = margin(t = -5, unit = "pt"),
                                                 axis.text.y = element_markdown(size=5, family = "sans"), # Rotate x-axis labels to 90 degrees
                                                  axis.text.x = element_markdown(size = 5, family = "sans")
)

p4 <- hm_combined + main_theme + theme(legend.position = "bottom", 
        legend.margin = margin(t = -3, b = 0, unit = "pt")) + guides(fill = guide_legend(title.position = "left", direction = "horizontal"))


p2_grob <- ggplotGrob(p2)
p3_grob <- ggplotGrob(p3)
p4_grob <- ggplotGrob(p4)
```


### Setting print layout for panel 1

```{r}
pdf("/home/bio.aau.dk/wy72yf/Projects/nanomotif-article/main_figures/figure_1_1.pdf", height = 5.51181102, width = 7.08661417)

layout1 <- grid.layout(nrow = 2, ncol = 2,
                      widths = unit(c(9, 11), "null"),
                      heights = unit(c(2, 1), "null"))


layout2 <- grid.layout(nrow = 2, ncol = 2,
                      widths = unit(c(9, 11), "null"),
                      heights = unit(c(18, 11), "null"), just = "top")

layout3 <- grid.layout(nrow = 2, ncol = 2,
                      widths = unit(c(9, 11), "null"),
                      heights = unit(c(17, 13), "null"), just = "left")

layout4 <- grid.layout(nrow = 2, ncol = 2,
                      widths = unit(c(9, 11), "null"),
                      heights = unit(c(19, 11), "null"))

# Push the layout to the drawing area
grid.newpage()

pushViewport(viewport(layout = layout2))

# For p3 in the top right area
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2))
grid.draw(p3_grob)
#grid.text("b", x = unit(0, "npc") + unit(4, "mm") , y = unit(1, "npc") - unit(3, "mm"), gp = gpar(fontfamily = "sans", fontface = "bold"))
upViewport()  # Go back up to the layout level

pushViewport(viewport(layout = layout4))

# For p2 in the bottom left area
pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 1))
grid.draw(p2_grob)
#grid.text("c", x = unit(0, "npc") + unit(4, "mm") , y = unit(1, "npc") - unit(2, "mm"), gp = gpar(fontfamily = "sans", fontface = "bold"))
upViewport()

pushViewport(viewport(layout = layout3))

# For p4 in the bottom right area
pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 2))
grid.draw(p4_grob)
#grid.text("d", x = unit(0, "npc") + unit(5, "mm") , y = unit(1, "npc") - unit(6, "mm"), gp = gpar(fontfamily = "sans", fontface = "bold"))
upViewport()

dev.off()


```



# Figure panel 2

### Loading plots for figure panel 1

```{r}
load("../RData/contamination_gc_cov_data.RData")
load("../RData/Contamination_heatmap.RData")
load("../RData/fecal_simple_hm_w_include.RData")
load("../RData/inclusion_gc_cov.RData")
```

```{r}
```

```{r}
### Adjusting theme and generating a grob object for each ggplot

main_theme <- theme(
    text = element_text(family = "sans", size = 1),
    plot.title = element_text(size = 7, family = "sans"),
    strip.text = element_markdown(size = 5, family = "sans"),
    strip.text.y.right = element_markdown(face = "bold", size = 5, family = "sans"),
    axis.text.y = element_markdown(size=5, family = "sans"), # Rotate x-axis labels to 90 degrees
    axis.text.x = element_markdown(size = 5, family = "sans"),
    axis.title.x = element_text(size = 7, family = "sans"),
    axis.title.y = element_text(size = 7, family = "sans"),
    axis.ticks = element_line(linewidth = 0.1),
    legend.title = element_text(size = 5, family = "sans"),
    legend.text = element_text(size = 5, family = "sans"),
    legend.key.size = unit(0.3, "cm"))


p1 <- contamination_gc_cov_plt & main_theme & theme(legend.position = "bottom") + theme(
                                                 legend.justification = c(1, 0),
                                                 legend.box.just = "right",
                                                 legend.margin = margin(t = -3, unit = "pt"))

p2 <- contamination_plt & main_theme
p2[[1]]$theme$axis.text.x <- element_blank()
p2[[2]]$theme$legend.margin = margin(t = -3, unit = "pt")

p3 <- include_plt_hm & main_theme & theme(axis.text.y = element_blank(),
                                          axis.text.x = element_blank())
p3[[1]]$theme$axis.text.y <- element_markdown(vjust = 0.5, hjust = 0, size = 5, family = "sans")

p3[[9]]$theme$axis.text.x <- element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 5, family = "sans")


p4 <- inclusion_gc_cov & main_theme + theme(legend.position = "bottom") & theme(
                                                 legend.justification = c(1, 0),
                                                 legend.box.just = "right",
                                                 legend.margin = margin(t = -3, unit = "pt"))


p5 <- p1 + p4 + plot_layout(guides = 'collect') & theme(legend.position = "bottom")

p1_grob <- as_grob(p1)
p2_grob <- as_grob(p2)
p3_grob <- as_grob(p3)
p4_grob <- as_grob(p4)
p5_grob <- as_grob(p5)


tb <- data.frame(
  row.names = c("# Contigs", "Contig N50\n(Kb)", "# MAGs\n(HQ)", "# Motifs\n(HQ)", "Motifs/MAG\n(HQ)", "%MAGs w/\nmotif (HQ)", "# MTases*", "# MTases in\nRM-systems*", "Linked motifs*\n(%motifs)", "Contamination\n(#contigs (HQ))", "Contigs included\n(#contigs (MGEs))"),
  Monocultures = c("-", "-", "10 (10)", "31 (31)", "3.1 (3.1)", "100% (100%)", "52", "21", "24 (77%)", "-", "-"),
  `Fecal(Simple)` = c("2034", "184.932", "14 (8)", "25 (19)", "1.8 (2.4)", "85% (100%)", "32", "9", "6 (32%)", "11 (0)", "135 (27)"),
  `Fecal(Complex)` = c("11625", "202.414", "190 (93)", "245 (192)", "1.3 (2.1)", "56% (75%)", "591", "247", "51 (27%)", "179 (20)", "712 (518)"),
    `Anerobic Digester` = c("65077", "98.395", "423 (230)", "667 (486)", "1.6 (2.1)", "60% (77%)", "1391", "583", "57 (12%)", "771 (121)", "3498 (744)"),
  `Soil` = c("374780", "51.739", "552 (66)", "173 (26)", "0.3 (0.4)", "30% (35%)", "105", "27", "8 (12%)", "435 (7)", "12472 (3399)"),
  `HQMAGs` = c("", "","", "", "", "", "", "","","", "") 
)


table_grob <- tableGrob(tb,
                        theme = ttheme_default(
                          core = list(fg_params = list(fontsize = 5, fontface = "plain", fontfamily = "sans")),
                          colhead = list(fg_params = list(fontsize = 5, fontface = "bold", fontfamily = "sans")),
                          rowhead = list(fg_params = list(fontsize = 5, fontface = "italic", fontfamily = "sans"))
                          #hsline = grid::gpar(col="grey", lwd=1),
                          #vsline = grid::gpar(col="grey", lwd=1),
                          #bgcolor = list(odd = "grey95", even = "grey90")
                        ))

ttb <- t(tb)

colnames(ttb) <- c("# Contigs", "Contig N50\n(Kb)", "# MAGs\n(HQ)", "# Motifs\n(HQ)", "Motifs/MAG\n(HQ)", "%MAGs w/\nmotif (HQ)", "# MTases*", "# MTases in\nRM-systems*", "Linked motifs*\n(%motifs)", "Contamination\n(#contigs (HQ))", "Contigs included\n(#contigs (MGEs))")

rownames(ttb) <- c("Monocultures", "Simple fecal", "Complex fecal", "Anaerobic digestor", "Soil", "*in HQ-MAGs")

tt1 <- ttheme_minimal(core = list(fg_params = list(fontsize = 5, fontface = "plain", fontfamily = "sans"), bg_params = list(fill = "white")),
                      colhead = list(fg_params = list(fontsize = 5, fontface = "bold", fontfamily = "sans"), border = TRUE),
                      rowhead = list(fg_params = list(fontsize = 5, fontface = "italic", fontfamily = "sans", hjust = 0, x = 0)))


ttable_grob <- tableGrob(ttb, theme = tt1) 


ttable_grob <- gtable_add_grob(ttable_grob,
        grobs = segmentsGrob( # line across the bottom
            x0 = unit(0,"npc"),
            y0 = unit(1,"npc"),
            x1 = unit(1,"npc"),
            y1 = unit(1,"npc"),
            gp = gpar(lwd = 2.0)),
        t = 2, b = 2, l = 1, r = ncol(ttable_grob))

ttable_grob <- gtable_add_grob(ttable_grob,
        grobs = segmentsGrob( # line across the bottom
            x0 = unit(0,"npc"),
            y0 = unit(1,"npc"),
            x1 = unit(1,"npc"),
            y1 = unit(1,"npc"),
            gp = gpar(lwd = 2.0)),
        t = 1, b = 1, l = 1, r = ncol(ttable_grob))

ttable_grob <- gtable_add_grob(ttable_grob,
        grobs = segmentsGrob( # line across the bottom
            x0 = unit(0,"npc"),
            y0 = unit(0,"npc"),
            x1 = unit(1,"npc"),
            y1 = unit(0,"npc"),
            gp = gpar(lwd = 2.0)),
        t = 6, b = 6, l = 1, r = ncol(ttable_grob))


#grid.newpage()

#grid.draw(ttable_grob)
```

```{r}
layout1 <- grid.layout(nrow = 2, ncol = 2,
                      widths = unit(c(1, 2), "null"),
                      heights = unit(c(1, 2), "null"))

layout2 <- grid.layout(nrow = 2, ncol = 2,
                      widths = unit(c(45, 60), "null"),
                      heights = unit(c(4, 1), "null"))

layout3 <- grid.layout(nrow = 3, ncol = 2,
                      widths = unit(c(1.09, 2.2), "null"),
                      heights = unit(c(1.18, 1.9, 0.72), "null"))

layout4 <- grid.layout(nrow = 4, ncol = 3,
                      widths = unit(c(1.4, 1.04, 1.06), "null"),
                      heights = unit(c(0.03, 1.09, 1.41, 1), "null"))

layout5 <- grid.layout(nrow = 4, ncol = 3,
                      widths = unit(c(1.4, 1.1, 1), "null"),
                      heights = unit(c(0.03, 1.09, 1.41, 1), "null"))

layout6 <- grid.layout(nrow = 2, ncol = 1,
                      widths = unit(c(1), "null"),
                      heights = unit(c(4.2, 1), "null"))


layout7 <- grid.layout(nrow = 4, ncol = 2,
                      widths = unit(c(1.02, 1.5), "null"),
                      heights = unit(c(0.00, 1.15, 1.41, 1), "null"))
```

```{r}
pdf("/home/bio.aau.dk/wy72yf/Projects/nanomotif-article/main_figures/figure_2.pdf", height = 7.4, width = 7.08661417, family = "sans")

grid.newpage()


pushViewport(viewport(layout = layout7))

pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 2))
grid.draw(p5_grob)
upViewport()




pushViewport(viewport(layout = layout3))

pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 2))
grid.draw(p3_grob)
upViewport()


pushViewport(viewport(layout = layout2))

# For p3 in the top right area
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1))
grid.draw(p2_grob)
#grid.text("b", x = unit(0, "npc") + unit(4, "mm") , y = unit(1, "npc") - unit(3, "mm"), gp = gpar(fontfamily = "sans", fontface = "bold"))
upViewport()

#pushViewport(viewport(layout = layout4))

#pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 2))
#grid.draw(p1_grob)
#upViewport()
 
#pushViewport(viewport(layout = layout5))

#pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 3))
#grid.draw(p4_grob)
#upViewport()


pushViewport(viewport(layout = layout6))

pushViewport(viewport(layout.pos.row = 2, layout.pos.col = 1))
grid.draw(ttable_grob)
upViewport()


dev.off()
```


```{r}
wrap_plots(mono_count_plot, sfecal_count_plot, 
           nrow = 2, ncol = 1, widths = c(1, 1))
```

```{r}
p1 <- ggplot(mtcars) + geom_point(aes(mpg, disp))
p2 <- ggplot(mtcars) + geom_boxplot(aes(gear, disp, group = gear))
p3 <- ggplot(mtcars) + geom_bar(aes(gear)) + facet_wrap(~cyl)

layout <- c(
  area(1, 1),
  area(1, 3, 3),
  area(3, 1, 3, 2)
)

# Show the layout to make sure it looks as it should
plot(layout)

# Apply it to a patchwork
p1 + p2 + p3 + plot_layout(design = layout)
```

```{r}

```

