---
title: "methylation_profile_embedding"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---



```{r}
pacman::p_load(
  "data.table",
  "ggplot2",
  "stringr",
  "tidyr",
  "dplyr",
  "ggtext",
  "pheatmap",
  "purrr",
  "here",
  "ggnewscale",
  "Biostrings",
  "umap"
)
getwd()
source("../src/constants.R")
source("../src/themes.R")
source("../src/utility.R")
```


```{r}
load_bin_motifs <- function(path){
  bin_motifs <- fread(path)
  return(bin_motifs)
}
load_motifs_scored <- function(path){
  motifs_scored <- fread(path)
  return(motifs_scored)
}
create_methylation_features <- function(bin_motifs, motifs_scored) {
  
}
```



```{r}
fecal_bin_motif <- load_bin_motif("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/fecal2/nanomotif/mmlong2/bin-motifs.tsv")
fecal_bin_motif[
  , moitf_uniq := paste0(motif, "_", mod_position, "_", mod_type)
]
bin_motifs_list <- fecal_bin_motif$moitf_uniq
fecal_motifs_scored <- load_motifs_scored("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/fecal2/nanomotif/mmlong2/motifs-scored.tsv")[
  , moitf_uniq := paste0(motif, "_", mod_position, "_", mod_type)
][moitf_uniq %in% bin_motifs_list]


dcast(contig, moitf_uniq, )
```
















