---
title: "benchmark"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---



```{r}
pacman::p_load(
  "data.table",
  "ggplot2",
  "stringr",
  "tidyr",
  "dplyr",
  "ggtext",
  "pheatmap",
  "purrr",
  "here",
  "ggnewscale",
  "Biostrings"
)
getwd()
source("../src/constants.R")
source("../src/themes.R")
source("../src/utility.R")
```


```{r}
benchmark_heatmap <- function(data) {
  hm <- ggplot(data) + 
  aes(x =n_motif_factor, y = coverage_factor) + 
  geom_tile(aes(fill = recall), color = "black") + 
  scale_fill_steps(limits = c(0, 1), low = "white",high = PLOT_COLORS[["dark_blue"]], na.value = "gray50", n.breaks = 6, labels = scales::percent) +
  facet_wrap(~motif_markdown, ncol = 2)+
  geom_text(aes(label=paste0(round(recall*100,   digits = 0), "%")), size=3) +
  theme_minimal() +
  theme(
    strip.text.x = element_markdown(),
    strip.text.y.right = element_markdown(face = "bold", size = 20)
  ) +
  coord_fixed(ratio= 1, clip="off") +
  labs(
    x = "Number of motif occurrences",
    y = "Read coverage",
    fill = "Motif\nRecall"
  )
  return(hm) 
}

extract_info <- function(filename) {
  list(
    sample = str_extract(filename, "(?<=benchmarks/)[^/]+(?=-contig)"),
    #contig = str_extract(filename, "contig_\\d+"),
    coverage = as.integer(str_extract(filename, "\\d+(?=x)")),
    #motif = str_extract(filename, "(?<=x\\.)[A-Za-z]+(?=\\d+)"),
    #pos = as.integer(str_extract(filename, "(?<=[A-Z])\\d+(?=[am]\\.)")),
    #mod_type = str_extract(filename, "[A-Za-z]+(?=\\.\\d+\\.motifs)"),
    n_motif = as.integer(str_extract(filename, "\\d+(?=\\.motifs)"))
  )
}
countMotifOccurrences <- function(fastaFilePath, motifRegex) {
  fasta <- readLines(fastaFilePath)
  
  as.character(readDNAStringSet(fastaFilePath)[[1]])
  sequences <- fasta[seq(2, length(fasta), 2)]
  totalOccurrences <- sum(sapply(sequences, function(seq) str_count(seq, motifRegex)))
  
  
  revCompSeq <- reverseComplement(fasta)
  revCompSeq <- revCompSeq[seq(2, length(revCompSeq), 2)]
  totalOccurrences <- sum(sapply(sequences, function(seq) str_count(seq, motifRegex))) + totalOccurrences
  
  return(totalOccurrences)
}
```


# Benchmarking of Nanomotif, intended usage

```{r}
folder_path <- "../data/nanomotif_benchmarks"

# List all files in the directory
files <- list.files(path = folder_path, recursive = TRUE, full.names = TRUE)

benchmark <- data.table(files)[
    , fread(files), by = files
  ][, c("sample", "coverage", "n_motif") := {
      info <- extract_info(files)
      .(info$sample, info$coverage, info$n_motif)
    }, by = files
  ][
    , coverage_factor := factor(coverage, levels = unique(sort(coverage)))
  ][
    , n_motif_factor := factor(n_motif, levels = unique(sort(n_motif, decreasing = TRUE)))
  ][
    , motif := unlist(lapply(motif, regex_to_iupac))
  ][
    , motif_markdown := paste0(
        str_sub(motif, 1, position),
        "<strong>",
        str_sub(motif, position+1, position+1),
        "<sub><sup>",
        map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
        "</sup></sub>",
        "</strong>",
        str_sub(motif, position+2, -1)
    )
  ]

```

## All plots

```{r}
plot_sample = "Geobacillus_3"
plot_data = benchmark[sample == plot_sample][
  , motif_markdown := factor(motif_markdown, levels = c("G<strong>A<sub><sup>m6</sup></sub></strong>TC", "GRNGA<strong>A<sub><sup>m6</sup></sub></strong>GY", "ACCC<strong>A<sub><sup>m6</sup></sub></strong>", "CCAA<strong>A<sub><sup>m6</sup></sub></strong>T"))
]
sample_title <- "*P. thermoglucosidasius*"
hm_geo <- benchmark_heatmap(plot_data) +
  labs(
    title = sample_title
  ) +
  theme(
    plot.title = element_markdown()
  )
plot_height <- 0.2 + (3 * ((length(unique(benchmark[sample == plot_sample]$motif)) - 1) %/% 2 +1))
ggsave("../figures/geobacillus_benchmark_heatmap.png", hm_geo, bg = "white", width = 7, height = plot_height)


plot_sample = "E.coli_K12_stat_3"
sample_title <- "*E. coli* K12"
hm_ec <- benchmark_heatmap(benchmark[sample == plot_sample][
  , motif_markdown := factor(motif_markdown, levels = c("G<strong>A<sub><sup>m6</sup></sub></strong>TC", "A<strong>A<sub><sup>m6</sup></sub></strong>CNNNNNNGTGC", "C<strong>C<sub><sup>m5</sup></sub></strong>WGG", "GC<strong>A<sub><sup>m6</sup></sub></strong>CNNNNNNGTT"))
]) +
  labs(
    title = sample_title
  ) +
  theme(
    plot.title = element_markdown()
  )
plot_height <- 0.2 + (3 * ((length(unique(benchmark[sample == plot_sample]$motif)) - 1) %/% 2 +1))
ggsave("../figures/ecoli_benchmark_heatmap.png", hm_ec, bg = "white", width = 7, height = plot_height)


plot_sample = "M.ruber_1"
sample_title <- "*M. ruber*"
hm_mr <- benchmark_heatmap(benchmark[sample ==plot_sample][
  , motif_markdown := factor(motif_markdown, levels = c("G<strong>A<sub><sup>m6</sup></sub></strong>TC", "GGG<strong>A<sub><sup>m6</sup></sub></strong>GC", "A<strong>A<sub><sup>m6</sup></sub></strong>TT", "TTA<strong>A<sub><sup>m6</sup></sub></strong>", "CC<strong>A<sub><sup>m6</sup></sub></strong>NNNNNNTGCC", "GGC<strong>A<sub><sup>m6</sup></sub></strong>NNNNNNTGG", "CTCG<strong>A<sub><sup>m6</sup></sub></strong>G"))
]) +
  labs(
    title = sample_title
  ) +
  theme(
    plot.title = element_markdown()
  )
plot_height <- 0.2 + (3 * ((length(unique(benchmark[sample == plot_sample]$motif)) - 1) %/% 2 +1))
ggsave("../figures/mruber_benchmark_heatmap.png", hm_mr, bg = "white", width = 7, height = plot_height)


# example plot
hm_ec
```

## Selected motif for main figure (Benchmark on microbemod pileup used instead)

```{r}
main_plot_data <- rbind(
  benchmark[sample == "E.coli_K12_stat_3"][motif == "GATC"],
  benchmark[sample == "Geobacillus_3"][motif == "GRNGAAGY"],
  benchmark[sample == "M.ruber_1"][motif == "GGCANNNNNNTGG"]
)
main_plot_data[
  , motif_markdown := paste0(
      str_sub(motif, 1, position),
      "<strong>",
      str_sub(motif, position+1, position+1),
      "<sub><sup>",
      map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
      "</sup></sub>",
      "</strong>",
      str_sub(motif, position+2, -1)
  )
]

hm <- ggplot(main_plot_data) + 
  aes(x =n_motif_factor, y = coverage_factor) + 
  geom_tile(aes(fill = recall), color = "black") + 
  scale_fill_steps(limits = c(0, 1), low = "white",high = PLOT_COLORS[["dark_blue"]], na.value = "gray50", n.breaks = 6) +
  facet_wrap(~motif_markdown, ncol = 3) + 
  geom_text(aes(label=paste0(round(recall*100,   digits = 0), "%"))) +
  theme_minimal() +
  theme(
    strip.text = element_markdown()
  ) +
  coord_fixed(ratio= 1, clip="off") +
  labs(
    x = "Number of motif occourances",
    y = "Read coverage"
  )

ggsave("../figures/select_benchmark_heatmap.png", hm, bg = "white", width = 10, height = 5)

```


# Microbemod only


```{r}

load_microbemod <- function(path) {
  ### Load microbemod output and select motif and convert information to be compatiable with benchmark parameters
  microbemod <- fread(path)[!is.na(Motif)][
    , complement_motif := unname(unlist(lapply(Motif, reverse_complement_iupac)))
  ][
    , mod_position := Methylated_position_1 - 1
  ][
    , mod_position_2 := Methylated_position_2 - 1
  ][
    , mod_type := unlist(lapply(Methylation_type, function(x) MOD_TYPE_SINGLE[x]))
  ][
    , motif_length := unlist(lapply(Motif, str_length))
  ][
    , complement_position := motif_length - Methylated_position_2
  ][, c("Motif", "mod_position", "mod_position_2", "complement_motif", "complement_position", "mod_type")][!is.na(complement_motif)]
  return(microbemod)
}
load_nanomotif <- function(path) {
  nanomotif <- fread(path)
  return(nanomotif)
}

check_motif_identified_microbemod <- function(path, motif, position, modification_type) {
  ### Check if motif,pos,type is identified in a microbemod output
  motif <- regex_to_iupac(motif)
  microbemod_output <- load_microbemod(path)
  microbemod_motif_match <- microbemod_output[
      (Motif == motif & mod_position == position & mod_type == modification_type) | 
      (Motif == motif & mod_position_2 == position & mod_type == modification_type) |
      (complement_motif == motif & complement_position == position & mod_type == modification_type)
    ]
  return(nrow(microbemod_motif_match) > 0 )
}
check_motif_identified_nanomotif <- function(path, motif_query, position, modification_type) {
  ### Check if motif,pos,type is identified in a microbemod output
  motif_query <- regex_to_iupac(motif_query)
  microbemod_output <- load_nanomotif(path)
  microbemod_motif_match <- microbemod_output[
      (motif == motif_query & mod_position == position & mod_type == modification_type) | 
      (motif_complement == motif_query & mod_position_complement == position & mod_type == modification_type)
        
    ]
  return(nrow(microbemod_motif_match) > 0 )
}

parse_benchmark_paths_to_datatable <- function(root_path, extension) {
  ### Identifies files in benchmark folder and returns the benchmark variables matching the regex pattern
  files <- list.files(path = root_path,pattern = extension,  full.names = TRUE, recursive = TRUE, include.dirs = TRUE)
  #.../{sample}-contig{contig}-nmotif{nmotif}-motif{motif}-type{mod_type}-pos{pos}-cov{coverage}/chunk_{chunk}
  pattern <- ".*/(.+)-contig(.+)-nmotif(.+)-motif(.+)-type(\\w+)-pos(\\d+)-cov(\\d+)/chunk_(\\d+)"
  
  # Initialize an empty data.table
  dt <- data.table(sample = character(), contig = character(), nmotif = integer(), 
                   motif = character(), mod_type = character(), pos = integer(), 
                   coverage = integer(), chunk = integer(), path = character())
  
  for (file in files) {
    matches <- regmatches(file, regexec(pattern, file))
    if (length(matches[[1]]) > 1) {
      dt <- rbind(
        dt, 
        data.table(
          sample = matches[[1]][2], contig = matches[[1]][3], 
          nmotif = matches[[1]][4], motif = matches[[1]][5], 
          mod_type = matches[[1]][6], pos = as.integer(matches[[1]][7]), 
          coverage = as.integer(matches[[1]][8]), chunk = as.integer(matches[[1]][9]),
          path = file
        )
      )
    }
  }
  return(dt)
}
parse_benchmark_paths_to_datatable_no_chunk <- function(root_path, extension) {
  ### Identifies files in benchmark folder and returns the benchmark variables matching the regex pattern
  files <- list.files(path = root_path,pattern = extension,  full.names = TRUE, recursive = TRUE, include.dirs = TRUE)
  #.../{sample}-contig{contig}-nmotif{nmotif}-motif{motif}-type{mod_type}-pos{pos}-cov{coverage}/chunk_{chunk}
  pattern <- ".*/(.+)-contig(.+)-nmotif(.+)-motif(.+)-type(\\w+)-pos(\\d+)-cov(\\d+)"
  
  # Initialize an empty data.table
  dt <- data.table(sample = character(), contig = character(), nmotif = integer(), 
                   motif = character(), mod_type = character(), pos = integer(), 
                   coverage = integer(), path = character())
  
  for (file in files) {
    matches <- regmatches(file, regexec(pattern, file))
    if (length(matches[[1]]) > 1) {
      dt <- rbind(
        dt, 
        data.table(
          sample = matches[[1]][2], contig = matches[[1]][3], 
          nmotif = as.integer(matches[[1]][4]), motif = matches[[1]][5], 
          mod_type = matches[[1]][6], pos = as.integer(matches[[1]][7]), 
          coverage = as.integer(matches[[1]][8]),
          path = file
        )
      )
    }
  }
  return(dt)
}
```

```{r}
full_count_bipartite_label <- fread("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/nanomotif_wf/M.ruber_1-contigcontig_1-nmotiffull-motifGGCA......TGG-typea-pos3-cov20/motifs.tsv")[motif == "GGCANNNNNNTGG"][, n := n_mod + n_nomod]$n
full_count_bipartite_label <- paste0(full_count_bipartite_label)
full_count_palindrome_label <- fread("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/nanomotif_wf/E.coli_K12_stat_3-contigcontig_3-nmotiffull-motifGATC-typea-pos1-cov20/motifs.tsv")[motif == "GATC"][, n := n_mod + n_nomod]$n
full_count_palindrome_label <- paste0(full_count_palindrome_label)
full_count_complex_label <- fread("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/nanomotif/Geobacillus_3-contigcontig_1-nmotiffull-motifG[AG].GAAG[CT]-typea-pos5-cov100/chunk_0/motifs.tsv")[motif == "GRNGAAGY"][, n := sum(n_mod) + sum(n_nomod)]$n[1]
full_count_complex_label <- paste0(full_count_complex_label)
full_count_generic_label = "All"
```


```{r}
mm_benchmark <- parse_benchmark_paths_to_datatable("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/microbemod", "motifs.tsv")
mm_mapping <- parse_benchmark_paths_to_datatable("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/mapping", "bam$")
mm_benchmark[
    , identified := unname(unlist(mapply(check_motif_identified_microbemod, path, motif, pos, mod_type)))
  ]
mm_summary <- mm_benchmark[
    , .(
      n_identified = sum(identified),
      n_ran  = .N
    ), by = .(motif, mod_type, pos, coverage, nmotif, sample)
  ]
mm_count <- mm_mapping[
    , .(n = .N),  by = .(motif, mod_type, pos, coverage, nmotif, sample)
  ]

mm_result <- merge(
    mm_summary, mm_count, all = TRUE
  )[
    , n_identified := fifelse(is.na(n_identified), 0, n_identified) 
  ][
    , recall := n_identified / n
  ][
    , coverage_factor := factor(coverage, levels = unique(sort(coverage)))
  ][
    , n_motif_factor := factor(nmotif, levels = c("All", full_count_bipartite_label, full_count_palindrome_label, full_count_complex_label, unique(sort(as.integer(nmotif), decreasing = TRUE))))
  ][
    , motif := unlist(lapply(motif, regex_to_iupac))
  ][
    , total_motif_count := fcase(
      motif == "GATC", full_count_palindrome_label,
      motif == "GGCANNNNNNTGG", full_count_bipartite_label,
      motif == "GRNGAAGY", full_count_complex_label
    )
  ][
    , motif_markdown := paste0(
        str_sub(motif, 1, pos),
        "<strong>",
        str_sub(motif, pos+1, pos+1),
        "<sub><sup>",
        map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
        "</sup></sub>",
        "</strong>",
        str_sub(motif, pos+2, -1)
    )
  ][
    , n_motif_factor := fcase(
      nmotif == "full", factor("All", levels = levels(n_motif_factor)),
      nmotif != "full" , n_motif_factor
    )
  ]

hm_mm <- ggplot(mm_result) + 
  aes(x =n_motif_factor, y = coverage_factor) + 
  geom_tile(aes(fill = recall), color = "black") + 
  scale_fill_steps(limits = c(0, 1), low = "white",high = PLOT_COLORS[["dark_blue"]], na.value = "gray50", n.breaks = 6) +
  facet_wrap(~motif_markdown, ncol = 3, scales = "free_x") + 
  geom_text(aes(label=paste0(round(recall*100,   digits = 0), "%"))) +
  theme_minimal() +
  theme(
    strip.text = element_markdown()
  ) +
  #coord_fixed(ratio= 1, clip="off") +
  labs(
    x = "Number of motif occourances",
    y = "Read coverage"
  )

ggsave("../figures/microbemod_benchmark_heatmap.png", hm_mm, bg = "white", width = 12, height = 4)
hm_mm
```


# Nanomotif on microbemod pileup

```{r}
nm_mm_benchmark <- parse_benchmark_paths_to_datatable("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/nanomotif/", "motifs-score-sub-noise-merge-complement.tsv")
nm_mm_benchmark[
  , identified := unname(unlist(mapply(check_motif_identified_nanomotif, path, motif, pos, mod_type)))
]
nm_mm_summary <- nm_mm_benchmark[
  , .(
    n_identified = sum(identified),
    n_ran  = .N
  ), by = .(motif, mod_type, pos, coverage, nmotif, sample)
]
nm_mm_count <- mm_mapping[
  , .(n = .N),  by = .(motif, mod_type, pos, coverage, nmotif, sample)
]

nm_mm_result <- merge(
  nm_mm_summary, nm_mm_count, all = TRUE
)[
    , n_identified := fifelse(is.na(n_identified), 0, n_identified) 
  ][
    , recall := n_identified / n
  ][
    , coverage_factor := factor(coverage, levels = unique(sort(coverage)))
  ][
    , n_motif_factor := factor(nmotif, levels = c("All", full_count_bipartite_label, full_count_palindrome_label, full_count_complex_label, unique(sort(as.integer(nmotif), decreasing = TRUE))))
  ][
    , motif := unlist(lapply(motif, regex_to_iupac))
  ][
    , total_motif_count := fcase(
      motif == "GATC", full_count_palindrome_label,
      motif == "GGCANNNNNNTGG", full_count_bipartite_label,
      motif == "GRNGAAGY", full_count_complex_label
    )
  ][
    , motif_markdown := paste0(
        str_sub(motif, 1, pos),
        "<strong>",
        str_sub(motif, pos+1, pos+1),
        "<sub><sup>",
        map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
        "</sup></sub>",
        "</strong>",
        str_sub(motif, pos+2, -1)
    )
  ][
    , n_motif_factor := fcase(
      nmotif == "full", factor("All", levels = levels(n_motif_factor)),
      nmotif != "full" , n_motif_factor
    )
  ]

hm_nm_mm <- ggplot(nm_mm_result) + 
  aes(x =n_motif_factor, y = coverage_factor) + 
  geom_tile(aes(fill = recall), color = "black") + 
  scale_fill_steps(limits = c(0, 1), low = "white",high = PLOT_COLORS[["dark_blue"]], na.value = "gray50", n.breaks = 6) +
  facet_wrap(~motif_markdown, ncol = 3) + 
  geom_text(aes(label=paste0(round(recall*100,   digits = 0), "%")), size = 1) +
  theme_minimal() +
  theme(
    strip.text = element_markdown()
  ) +
  coord_fixed(ratio= 1, clip="off") +
  labs(
    x = "Number of motif occourances",
    y = "Read coverage"
  )
hm_nm_mm
```
# Comparison of benchmark

```{r}
combined_benchmark <- rbind(
mm_result[, c("motif", "motif_markdown", "recall", "coverage_factor", "n_motif_factor", "total_motif_count")][, tool := "MicrobeMod"],
nm_mm_result[, c("motif","motif_markdown", "recall", "coverage_factor", "n_motif_factor", "total_motif_count")][, tool := "Nanomotif"]
)
combined_benchmark[
  , motif_markdown := case_when(
    motif == "GATC" ~ paste0("*E. coli* K12<br>",  motif_markdown, "<br> Palindromic<br>All=", total_motif_count),
    motif == "GRNGAAGY" ~ paste0("*P. thermoglucosidasius*<br>", motif_markdown, "<br>Short Non-Palindromic<br>All=", total_motif_count),
    motif == "GGCANNNNNNTGG" ~ paste0("*M. ruber*<br>", motif_markdown, "<br> Bipartite<br>All=", total_motif_count)
    
  )
][
  , tool := factor(tool, levels = c("Nanomotif", "MicrobeMod"))
]
hm_combined <- ggplot(combined_benchmark) + 
  aes(x =n_motif_factor, y = coverage_factor) + 
  geom_tile(aes(fill = recall), color = "black") + 
  scale_fill_steps(limits = c(0, 1), low = "white",high = PLOT_COLORS[["dark_blue"]], na.value = "gray50", n.breaks = 6, labels = scales::percent) +
  facet_grid(tool ~motif_markdown, scale = "free_x")+
  geom_text(aes(label=paste0(round(recall*100,   digits = 0), "%")), size=1) +
  theme_minimal() +
  theme(
    strip.text.x = element_markdown(),
    strip.text.y.right = element_markdown(face = "bold", size = 20)
  ) +
  #coord_fixed(ratio= 1, clip="off") +
  labs(
    x = "Number of motif occurrences",
    y = "Read coverage",
    fill = "Motif\nRecall"
  )
ggsave("../figures/benchmark_heatmap_both.png", hm_combined, bg = "white", width = 10, height = 6)
save(hm_combined, file = "../figures/benchmark_heatmap_both.Rdata")
hm_combined
```


# Bin consensus
```{r}
nm_bin_benchmark <- parse_benchmark_paths_to_datatable_no_chunk("/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/microbemod_benchmark/nanomotif_wf", "bin-motifs.tsv")[!is.na(nmotif)]
nm_bin_benchmark[
  , motif := fifelse(
    motif == "GAG.GAAGCT",
    "G[AG].GAAG[CT]",
    motif
  )
][
  , identified := unname(unlist(mapply(check_motif_identified_nanomotif, path, motif, pos, mod_type)))
][
  , coverage_factor := factor(coverage, levels = unique(sort(coverage)))
][
  , n_motif_factor := factor(nmotif, levels = unique(sort(nmotif, decreasing = TRUE)))
][
  , motif := fifelse(
    motif == "GAG.GAAGCT",
    "G[AG].GAAG[CT]",
    motif
  )
][
  , motif := unlist(lapply(motif, regex_to_iupac))
][
  , motif_markdown := paste0(
      str_sub(motif, 1, pos),
      "<strong>",
      str_sub(motif, pos+1, pos+1),
      "<sub><sup>",
      map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
      "</sup></sub>",
      "</strong>",
      str_sub(motif, pos+2, -1)
  )
][
  , motif_markdown := case_when(
    motif == "GATC" ~ paste0("*E. coli* K12<br>",  motif_markdown, "<br> Palindromic"),
    motif == "GRNGAAGY" ~ paste0("*P. thermoglucosidasius*<br>", motif_markdown, "<br> Short Non-Palindromic"),
    motif == "GGCANNNNNNTGG" ~ paste0("*M. ruber*<br>", motif_markdown, "<br> Bipartite")
    
  )
]
bin_consensus_benchmark <- ggplot(nm_bin_benchmark) + 
  aes(x =n_motif_factor, y = coverage_factor) + 
  geom_tile(aes(fill = identified), color = "black", linewidth =0.1) + 
  facet_wrap(~motif_markdown, ncol = 3) + 
  theme_minimal() +
  theme(
    strip.text = element_markdown()
  ) +
  coord_fixed(ratio= 1, clip="off") +
  labs(
    x = "Number of motif occurrences",
    y = "Read coverage",
    fill = "Identified in\nbin-consensus"
  ) +
  scale_fill_manual(values = c("TRUE" = "#3A481E", "FALSE"="gray80"))
bin_consensus_benchmark
ggsave("../figures/benchmark_heatmap_bin_consensus.png", bin_consensus_benchmark, bg = "white", width = 10, height = 3.5)
save(bin_consensus_benchmark, file = "../figures/benchmark_heatmap_bin_consensus.Rdata")
```





```{r}
bin_consensus_benchmark
```


```{r}
```


# Full genome benchmark

```{r}
fasta <- "/home/bio.aau.dk/lx38ll/dark-science/motif-identification/analysis/Geobacillus_3/medaka/assembly.polished.fasta"


as.character(readDNAStringSet(fasta))
```

