---
title: "R Notebook"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
editor_options:
  chunk_output_type: inline
---


```{r}
pacman::p_load(
  "data.table",
  "ggplot2",
  "stringr",
  "tidyr",
  "dplyr",
  "ggtext",
  "pheatmap",
  "purrr",
  "here",
  "ggnewscale",
  "gt"
)
setwd(here::here())
source("src/constants.R")
source("src/themes.R")
source("src/utility.R")
direct_label = "Direct"
indirect_label = "Indirect"
```


```{r}
load_complex_sample <- function(
    motifs_direct_path,
    motifs_indifrect_path,
    motifs_bin_path,
    contig_bin_path,
    bin_info_path
  ){
  motifs_direct <-  fread(motifs_direct_path)[, directly := TRUE]
  motifs_score <- fread(motifs_indifrect_path)
  motifs_full <- merge(
    motifs_score, motifs_direct, all.x = TRUE, by=c("contig", "motif", "mod_position", "mod_type"), suffixes = c("", ".drop"))[
    , directly := fifelse(is.na(directly), indirect_label, direct_label)
  ]
  
  bin_info <-  fread(bin_info_path)
  if("contamination_checkm2" %in% colnames(bin_info)){
    bin_info[
      , bin_quality := case_when(
        completeness_checkm2 > 90 & contamination_checkm2 < 5 ~ "HQ",
        completeness_checkm2 > 50 & contamination_checkm2 < 10 ~ "MQ",
        TRUE ~ "LQ"
      )
    ]
  } else {
      bin_info[
      , bin_quality := case_when(
        Completeness > 90 & Contamination < 5 ~ "HQ",
        Completeness > 50 & Contamination < 10 ~ "MQ",
        TRUE ~ "LQ"
      )
    ]
  }
  
  bin <- fread(contig_bin_path, header = FALSE)
  colnames(bin) <-  c("contig", "bin")
  bin[, n_contigs := .N, by = bin]
  
  bin_motifs <- fread(motifs_bin_path) 
  
  motifs_full <- merge(motifs_full, bin, by = "contig", all.x = TRUE) 
  
  
  motifs_full <- motifs_full[motif %in% bin_motifs$motif][
      , mean := n_mod / (n_mod + n_nomod-1)
    ][
        , name := contig
    ][
      , motif_markdown := paste0(
          str_sub(motif, 1, mod_position),
          "**",
          map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
          str_sub(motif, mod_position+1, mod_position+1),
          "**",
          str_sub(motif, mod_position+2, -1)
      )
    ][
      , simple_count := fcase(
        (n_mod + n_nomod) > 10, "",
        (n_mod + n_nomod) <= 10, as.character(n_mod + n_nomod - 1)
      )
    ]
  
  
  bin_motifs[
      , mean := n_mod_bin / (n_mod_bin + n_nomod_bin-1)
    ][
      , name := bin
    ][
      , motif_markdown := paste0(
          str_sub(motif, 1, mod_position),
          "**",
          map(mod_type, function(x) MOD_TYPE_PRETTY[[x]]) %>%  unlist(),
          str_sub(motif, mod_position+1, mod_position+1),
          "**",
          str_sub(motif, mod_position+2, -1)
      )
    ]
  return(
    list(
      "bin_info"=bin_info,
      "bin_motifs"=bin_motifs,
      "motifs_full"=motifs_full
    )
  )
}

load_sample <- function(sample){
  sample_data <- load_complex_sample(
    paste0("../data/real_communities/", sample, "/nanomotif_0.4.13/motifs.tsv"),
    paste0("../data/real_communities/", sample, "/nanomotif_0.4.13/motifs-scored.tsv"),
    paste0("../data/real_communities/", sample, "/nanomotif_0.4.13/bin-motifs.tsv"),
    paste0("../data/real_communities/", sample, "/mmlong2_lite/tmp/binning/contig_bin.tsv"),
    paste0("../data/real_communities/", sample, "/mmlong2_lite/results/mmlong2_lite_bins.tsv")
  )
  return(sample_data)
}

calculate_sample_stats <- function(sample_data, sample_name){
  sample_hq_bin = sample_data[["bin_info"]][bin_quality == "HQ"]$bin
  sample_stats <- data.table(
    sample = sample_name,
    n_bin = nrow(sample_data[["bin_info"]]),
    n_hq = nrow(sample_data[["bin_info"]][bin_quality == "HQ"]),
    n_mq = nrow(sample_data[["bin_info"]][bin_quality == "MQ"]),
    n_bin_with_motif = length(unique(sample_data[["bin_motifs"]]$bin)),
    n_hq_with_motif = sum(unique(sample_data[["bin_motifs"]]$bin) %in% sample_hq_bin),
    n_motif = nrow(sample_data[["bin_motifs"]][mean > 0.5]),
    n_5mC_motifs = nrow(sample_data[["bin_motifs"]][mean > 0.5][mod_type == "m"]),
    n_6mA_motifs = nrow(sample_data[["bin_motifs"]][mean > 0.5][mod_type == "a"]),
    n_4mC_motifs = nrow(sample_data[["bin_motifs"]][mean > 0.5][mod_type == "21839"]),
    n_motif_hq = nrow(sample_data[["bin_motifs"]][bin %in% sample_hq_bin][mean > 0.5])
  )
  sample_stats[
    , motif_pr_bin := n_motif / n_bin
  ][
    , motif_pr_bin_hq := n_motif_hq / n_hq
  ]
  return(sample_stats)
}

get_axis_order <- function(wide_df) {
  set.seed(1)
  feature_matrix = as.matrix(wide_df[,-1] )
  y_row_order <- pheatmap::pheatmap(as.matrix(feature_matrix), silent = TRUE)$tree_row$order
  y_factor_order <- wide_df$name[y_row_order]
  x_row_order <- pheatmap::pheatmap(t(as.matrix(feature_matrix)), silent = TRUE)$tree_row$order
  x_factor_order <- colnames(wide_df[,-1])[x_row_order]
  return(list(
    y = y_factor_order,
    x = x_factor_order
  ))
}
```


```{r}
plot_faceted_motif_heatmap <- function(sample_data) {
  plot_data <- sample_data[["motifs_full"]][
    , any_direct := "Direct" %in% directly, by = bin
  ]
  
  plot_data[
    , bin := str_extract(bin, "bin.*")
  ]
  
  axis_orders <- get_axis_order(dcast(plot_data, name ~ motif_markdown, value.var = "mean", fill = 0, fun.aggregate = mean))
  bin_contigs_heatmap <- ggplot(plot_data) +
    aes(x = factor(motif_markdown, levels = axis_orders[["x"]]), y = factor(name, levels =  axis_orders[["y"]])) +
    geom_tile(aes(fill = directly)) +
    scale_fill_manual(values = c("Direct"=PLOT_COLORS[[1]], "Indirect"=PLOT_COLORS[[2]])) +
    guides(fill = guide_legend(
      title = "Method of detection",
      override.aes = list(alpha = 1))
    ) +
    
    new_scale("fill") +
    geom_tile(aes(fill = mean),  color = "#ffffff00",linewidth = 0) +
    scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "red", high = PLOT_COLORS[[1]], na.value = "white") +
    
    new_scale("fill") +
    geom_tile(data=plot_data[directly == indirect_label], aes(fill = mean),  color = "#ffffff00",linewidth = 0) + 
    scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "red", high = PLOT_COLORS[[2]], na.value = "white") +
    labs(fill = "Directly") +  
    
    new_scale("fill") +
    geom_text(data=plot_data[(n_mod + n_nomod - 1)==0], label = ".", hjust=0.5, vjust=0, size=1) +
    geom_tile(aes(fill = mean), linewidth = 0, alpha = 0, color = "#33333300") +
    scale_fill_gradient2(
      limits = c(0, 1), low = "red", high = "gray20",
      labels = scales::percent
    ) +
    
    theme(
      axis.text.y = element_blank(),# element_markdown(vjust = 0.5, hjust = 0, size = 0.5),
      axis.ticks.y = element_blank(),
      axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 5),
      panel.grid.minor = element_blank(),  # Remove minor gridlines
      panel.background = element_blank(),   # Remove panel background,
      panel.margin = unit(0, "lines"),
      panel.border = element_rect(colour = "#141414", fill=NA, linewidth = 0.1),
      strip.placement = "outside",
      strip.text.y.left = element_markdown(angle = 0, vjust = 0.5, hjust = 0),
      strip.background = element_blank()
    ) +
    labs(
      x = "Motif",
      y = "",
      fill = "Methylation degree"
    ) +
    facet_wrap(~bin, scales = "free_y", ncol = 1, strip.position = "left")+
    geom_point(aes(color = "No motif observations"), shape = 16, size = 1, alpha = 0) +
    scale_color_manual(name = "Indicator", values = c("No motif observations" = "black")) +
    guides(color = guide_legend(override.aes = list(alpha = 1, shape = c(16)))) # Adjust this line as needed
  return(bin_contigs_heatmap)
}
```


```{r}
fecal_simple <- load_sample("fecal_simple")
zymo_fecal <- load_sample("ZymoFecal")
fecal_complex <- load_sample("fecal_inhouse")
ad <- load_sample("anaerobic_digester")
soil <- load_sample("mfd02199_backup")

```

# Summary numbers

```{r}
fecal_simple_stats <- calculate_sample_stats(fecal_simple, "Fecal Simple")
fecal_complex_stats <- calculate_sample_stats(fecal_complex, "Fecal Complex")
zymo_fecal_stats <- calculate_sample_stats(zymo_fecal, "ZymoFecal")
ad_stats <-   calculate_sample_stats(ad, "Anaerobic digester")
soil_stats <- calculate_sample_stats(soil, "Soil")
sample_summary <- rbind(
  fecal_simple_stats,
  zymo_fecal_stats,
  fecal_complex_stats,
  ad_stats,
  soil_stats
)
summary_table <- sample_summary %>%  gt() %>%  
  tab_style(
    style = cell_borders(sides = "all", color = "#000000", style = "solid"),
    
    locations = cells_body()
  )
```

# PLOTS

```{r}
zymo_fecal_plot_faceted_motifs_heatmap <- plot_faceted_motif_heatmap(zymo_fecal)
height_hm <- 9
width_hm <- 16
ggsave(
  paste0("../figures/motif_heatmap_faceted_on_bins/ZymoFecal.png"),
  zymo_fecal_plot_faceted_motifs_heatmap,
  height = height_hm,
  width = width_hm,
  dpi = "retina",
  bg = "white")
```
```{r}
fecal_simple_plot_faceted_motifs_heatmap <- plot_faceted_motif_heatmap(fecal_simple)
height_hm <- 9
width_hm <- 16
ggsave(
  paste0("../figures/motif_heatmap_faceted_on_bins/fecal_simple.png"),
  fecal_simple_plot_faceted_motifs_heatmap,
  height = height_hm,
  width = width_hm,
  dpi = "retina",
  bg = "white")
```

```{r}
asdf = fecal_simple$motifs_full
ggplot(fecal_simple$motifs_full) +
  aes(x = motif, y = contig, fill = mean) +
  geom_tile() +
  facet_wrap(~bin, scales = "free_y", ncol = 1, strip.position = "left")
```



```{r}

axis_orders <- get_axis_order(dcast(motifs_full_binned, name ~ motif_markdown, value.var = "mean", fill = 0, fun.aggregate = mean))
hm <- ggplot(motifs_full_binned) +
  aes(x = factor(motif_markdown, levels = axis_orders[["x"]]), y = factor(name, levels = axis_orders[["y"]])) + 
  geom_tile(aes(fill = directly), linewidth = 0, alpha = 0) + 
  scale_fill_manual(values = c("Direct"="forestgreen", "Indirect"="#8C228C")) +
  guides(fill = guide_legend(
    title = "Method of detection",
    override.aes = list(alpha = 1))
  ) +
  new_scale("fill") +
  geom_tile(aes(fill = mean), linewidth = 0) +   
  scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "white", high = "forestgreen", na.value = "white") +
  new_scale("fill") +
  geom_tile(data=motifs_full_binned[directly == indirect_label], aes(fill = mean), linewidth = 0) + 
  scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "white", high = "#8C228C", na.value = "white") +
  labs(fill = "Directly") + 
  new_scale("fill") +
  geom_tile(aes(fill = mean), linewidth = 0, color = "gray20", alpha = 0) +
  scale_fill_gradient2(
    limits = c(0, 1), low = "white", high = "gray20",
    labels = scales::percent
  ) +
  #geom_text(data=fecal_binned[(n_mod + n_nomod - 1)==0], label = ".", hjust=0.5, vjust=0, size=0.1) +
  #coord_fixed(ratio= 1, clip="off") +
  theme(
    axis.text.y = element_markdown(vjust = 0.5, hjust = 0, size = 1), # Rotate x-axis labels to 90 degrees
    axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 2),
    panel.grid.minor = element_blank(),  # Remove minor gridlines
    panel.background = element_blank(),   # Remove panel background,
  ) +
  labs(
    x = "Motif",
    y = "",
    fill = "Methylation degree"
  )
```



```{r}
ggsave(
  "../figures/binned_contig_heatmap_", sample, ".png",
  hm,
  width = 25,
  height = 25
)
```

```{r}

axis_orders <- get_axis_order(dcast(bin_motifs, name ~ motif_markdown, value.var = "mean", fill = 0, fun.aggregate = mean))
hm_fecal <- ggplot(bin_motifs) +
  aes(x = factor(motif_markdown, levels = axis_orders[["x"]]), y = factor(name, levels = axis_orders[["y"]])) + 
  geom_tile(aes(fill = mean), linewidth = 0) +   
  scale_fill_gradient2(guide = "none", limits = c(0, 1), low = "white", high = "forestgreen", na.value = "white") +
  new_scale("fill") +
  geom_tile(aes(fill = mean), linewidth = 0, color = "gray20", alpha = 0) +
  scale_fill_gradient2(
    limits = c(0, 1), low = "white", high = "gray20",
    labels = scales::percent
  ) +
  #geom_text(data=bin_motifsned[(n_mod + n_nomod - 1)==0], label = ".", hjust=0.5, vjust=0, size=0.1) +
  #coord_fixed(ratio= 1, clip="off") +
  theme(
    axis.text.y = element_markdown(vjust = 0.5, hjust = 0, size = 1), # Rotate x-axis labels to 90 degrees
    axis.text.x = element_markdown(angle = 90, hjust = 1, vjust = 0.5, size = 2),
    panel.grid.minor = element_blank(),  # Remove minor gridlines
    panel.background = element_blank(),   # Remove panel background,
  ) +
  labs(
    x = "Motif",
    y = "",
    fill = "Methylation degree"
  )
hm_fecal
```

```{r}
bin_motifs[
  ,  c("domain", "phylum", "class", "order", "family", "genus", "species") := extraxt_gtdb_tax(classification)
]
```





```{r}
ggplot(bin_motifs) + 
  geom_bar(aes(x=mod_type, fill = motif_type), color = "gray20") +
  base_plot_theme() +
  scale_fill_manual(values = c("ambiguous"="#F16A70", "bipartite"="#B1D877","non-palindrome"= "#8CDCDA","palindrome"= "#4D4D4D"))
  
```

```{r}
motifs_summary <-  bin_motifs[mean > 0.5][, .(
    n_motifs = .N
  ), by = .(motif, mod_position, mod_type, motif_type)][
    , group := factor(case_when(
      sum(n_motifs) == 1 ~ "1",
      sum(n_motifs) > 1 & sum(n_motifs) < 5 ~ "2-4",
      sum(n_motifs) >= 5 & sum(n_motifs) < 10 ~ "5-9",
      sum(n_motifs) >= 10 ~ "10+"
    ), levels = c("1", "2-4", "5-9", "10+")), by = .(motif, mod_position, mod_type)
  ]
```


```{r}
ggplot(motifs_summary[group != "1"]) + 
  geom_histogram(aes(x = group, fill = motif_type), stat="count") + 
  base_plot_theme() + 
  facet_wrap(~mod_type) +
  scale_fill_manual(values = c("ambiguous"="#f1a86a", "non-palindrome"="#a3d15e","bipartite"= "#8cb6dc","palindrome"= "#4D4D4D")) +
  ylim(0, NA)
```


```{r}
ggplot(bin_motifs) +
  geom_histogram(aes(x=mean)) +
  xlim(0, 1) 
  
```


```{r}
bin_motifs
```
















